<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell type annotation diagnostics • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Cell type annotation diagnostics">
<meta property="og:description" content="The scDiagnostics package provides diagnostic plots to assess the quality of cell type assignments from single cell gene expression profiles. The implemented functionality allows to assess the reliability of cell type annotations, investigate gene expression patterns, and explore relationships between different cell types in query and reference datasets allowing users to detect potential misalignments between reference and query datasets. The package also provides visualization capabilities for diagnositics purposes.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/scDiagnostics.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="scdiagnostics-diagnostic-functions-to-assess-the-quality-of-cell-type-annotations-in-single-cell-rna-seq-data">scDiagnostics: diagnostic functions to assess the quality of cell type annotations in single-cell RNA-seq data<a class="anchor" aria-label="anchor" href="#scdiagnostics-diagnostic-functions-to-assess-the-quality-of-cell-type-annotations-in-single-cell-rna-seq-data"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h1>
<p>Annotation transfer from a reference dataset for the cell type annotation of a new query single-cell RNA-sequencing (scRNA-seq) experiment is an integral component of the typical analysis workflow. The approach provides a fast, automated, and reproducible alternative to the manual annotation of cell clusters based on marker gene expression. However, dataset imbalance and undiagnosed incompatibilities between query and reference dataset can lead to erroneous annotation and distort downstream applications.</p>
<p>The <code>scDiagnostics</code> package provides functionality for the systematic evaluation of cell type assignments in scRNA-seq data. <code>scDiagnostics</code> offers a suite of diagnostic functions to assess whether both (query and reference) datasets are aligned, ensuring that annotations can be transferred reliably. <code>scDiagnostics</code> also provides functionality to assess annotation ambiguity, cluster heterogeneity, and marker gene alignment. The implemented functionality helps researchers to determine how accurately cells from a new scRNA-seq experiment can be assigned to known cell types.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>To install the development version of the package from GitHub use the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: you will need the <a href="https://cran.r-project.org/web/packages/remotes/index.html" class="external-link">remotes</a> package to install from GitHub.</p>
<p>To build the package vignettes upon installation use:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span>,</span>
<span>                     build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h1>
<p>To explore the capabilities of the scDiagnostics package, you can load your own data or utilize publicly available datasets obtained from the scRNAseq R package. In this guide, we will demonstrate how to use scDiagnostics with such datasets, which serve as valuable resources for exploring the package and assessing the appropriateness of cell type assignments.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Loading libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://scenic.aertslab.org" class="external-link">AUCell</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex" class="external-link">celldex</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="exploratory-analysis-qc-stats-vs-annotation-scores">Exploratory Analysis: QC Stats vs. Annotation Scores<a class="anchor" aria-label="anchor" href="#exploratory-analysis-qc-stats-vs-annotation-scores"></a>
</h2>
<p>Here, we will consider the Human Primary Cell Atlas (Mabbott et al. 2013) as a reference dataset and our query dataset consists of Haematopoietic stem and progenitor cells from (Bunis DG et al. 2021).</p>
<p>In scRNA-seq studies, assessing the quality of cells is important for accurate downstream analyses. At the same time, assigning accurate cell type labels based on gene expression profiles is an integral aspect of scRNA-seq data interpretation. Generally, these two are performed independently of each other. The rationale behind this function is to inspect whether certain QC (Quality Control) criteria impact the confidence level of cell type annotations.</p>
<p>For instance, it is reasonable to hypothesize that higher library sizes could contribute to increased annotation confidence due to enhanced statistical power for identifying cell type-specific gene expression patterns, as evident in the scatter plot below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># load reference dataset</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/fetchReference.html" class="external-link">fetchReference</a></span><span class="op">(</span><span class="st">"hpca"</span>, <span class="st">"2024-02-26"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load query dataset (Bunis haematopoietic stem and progenitor cell data) from </span></span>
<span><span class="co"># Bunis DG et al. (2021). Single-Cell Mapping of Progressive Fetal-to-Adult </span></span>
<span><span class="co"># Transition in Human Naive T Cells Cell Rep. 34(1): 108573</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/BunisHSPCData.html" class="external-link">BunisHSPCData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span></span>
<span></span>
<span><span class="co"># Add QC metrics to query data</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Log transform query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run SingleR to predict cell types</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign predicted labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Get annotation scores</span></span>
<span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">scores</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">cell_scores</span> <span class="op">&lt;-</span> <span class="va">scores</span></span>
<span></span>
<span><span class="co"># Create a scatter plot between library size and annotation scores</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    qc_col <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>    label_col <span class="op">=</span> <span class="st">"pred.labels"</span>,</span>
<span>    score_col <span class="op">=</span> <span class="st">"cell_scores"</span>,</span>
<span>    label <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Scatter-Plot-LibrarySize-Vs-Annotation-Scores-1.png" width="100%"></p>
<p>However, certain QC metrics, such as the proportion of mitochondrial genes, may require careful consideration as they can sometimes be associated with cellular states or functions rather than noise. The interpretation of mitochondrial content should be context-specific and informed by biological knowledge.</p>
<p>In next analysis, we investigated the relationship between mitochondrial percentage and cell type annotation scores using liver tissue data from He S et al. 2020. Notably, we observed high annotation scores for macrophages and monocytes. These findings align with the established biological characteristic of high mitochondrial activity in macrophages and monocytes, adding biological context to our results.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HeOrganAtlasData.html" class="external-link">HeOrganAtlasData</a></span><span class="op">(</span></span>
<span>    tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Liver"</span><span class="op">)</span>,</span>
<span>    ensembl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    location <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add QC metrics to query data</span></span>
<span><span class="va">mito_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">unfiltered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">query_data</span>,subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mt <span class="op">=</span> <span class="va">mito_genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/quickPerCellQC.html" class="external-link">quickPerCellQC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span>,sub.fields<span class="op">=</span><span class="st">"subsets_mt_percent"</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>,<span class="op">!</span><span class="va">qc</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Log transform query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run SingleR to predict cell types</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign predicted labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Calculate cell scores</span></span>
<span><span class="co"># Get annotation scores</span></span>
<span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">scores</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">cell_scores</span> <span class="op">&lt;-</span> <span class="va">scores</span></span>
<span></span>
<span><span class="co"># Create a new column for the labels so it is easy to distinguish</span></span>
<span><span class="co">#  between Macrophoges, Monocytes and other cells</span></span>
<span><span class="va">query_data</span><span class="op">$</span><span class="va">label_category</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>, <span class="st">"Monocyte"</span><span class="op">)</span>,</span>
<span>                                     <span class="va">query_data</span><span class="op">$</span><span class="va">pred.labels</span>,</span>
<span>                                     <span class="st">"Other cells"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Define custom colors for cell type labels</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Other cells"</span> <span class="op">=</span> <span class="st">"grey"</span>, <span class="st">"Macrophage"</span> <span class="op">=</span> <span class="st">"green"</span>, <span class="st">"Monocyte"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate scatter plot for all cell types</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    qc_col <span class="op">=</span> <span class="st">"subsets_mt_percent"</span>,</span>
<span>    label_col <span class="op">=</span> <span class="st">"label_category"</span>,</span>
<span>    score_col <span class="op">=</span> <span class="st">"cell_scores"</span>,</span>
<span>    label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"subsets_mt_percent"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="reference/figures/QC-Annotation-Scatter-Mito-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="examining-distribution-of-qc-stats-and-annotation-scores">Examining Distribution of QC stats and Annotation Scores<a class="anchor" aria-label="anchor" href="#examining-distribution-of-qc-stats-and-annotation-scores"></a>
</h2>
<p>In addition to the scatter plot, we can gain further insights into the gene expression profiles by visualizing the distribution of user defined QC stats and annotation scores for all the cell types or specific cell types. This allows us to examine the variation and patterns in expression levels and scores across cells assigned to the cell type of interest.</p>
<p>To accomplish this, we create two separate histograms. The first histogram displays the distribution of the annotation scores.</p>
<p>The second histogram visualizes the distribution of QC stats. This provides insights into the overall gene expression levels for the specific cell type. Here in this particular example we are investigating percentage of mitochondrial genes.</p>
<p>By examining the histograms, we can observe the range, shape, and potential outliers in the distribution of both annotation scores and QC stats. This allows us to assess the appropriateness of the cell type assignments and identify any potential discrepancies or patterns in the gene expression profiles for the specific cell type.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate histogram</span></span>
<span><span class="fu"><a href="reference/histQCvsAnnotation.html">histQCvsAnnotation</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, qc_col <span class="op">=</span> <span class="st">"subsets_mt_percent"</span>, </span>
<span>                   label_col <span class="op">=</span> <span class="st">"pred.labels"</span>, </span>
<span>                   score_col <span class="op">=</span> <span class="st">"cell_scores"</span>, </span>
<span>                   label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Mito-Genes-Vs-Annotation-1.png" width="100%"></p>
<p>The right-skewed distribution for mitochondrial percentages and a left-skewed distribution for annotation scores in above histograms suggest that most cells have lower mitochondrial contamination and higher confidence in their assigned cell types.</p>
</div>
<div class="section level2">
<h2 id="exploring-gene-expression-distribution">Exploring Gene Expression Distribution<a class="anchor" aria-label="anchor" href="#exploring-gene-expression-distribution"></a>
</h2>
<p>This function helps user to explore the distribution of gene expression values for a specific gene of interest across all the cells in both reference and query datasets and within specific cell types. This helps to evaluate whether the distributions are similar or aligned between the datasets. Discrepancies in distribution patterns may indicate potential incompatibilities or differences between the datasets.</p>
<p>The function also allows users to narrow down their analysis to specific cell types of interest. This enables investigation of whether alignment between the query and reference datasets is consistent not only at a global level but also within specific cell types.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load data</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HeOrganAtlasData.html" class="external-link">HeOrganAtlasData</a></span><span class="op">(</span>tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Marrow"</span><span class="op">)</span>, ensembl <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Divide the data into reference and query datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.7</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">indices</span><span class="op">]</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">-</span><span class="va">indices</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Log-transform datasets</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ref_data</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run PCA</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get cell type scores using SingleR</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">reclustered.broad</span><span class="op">)</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pred</span><span class="op">)</span></span>
<span>   </span>
<span><span class="co"># Assign labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span>   </span>
<span><span class="co"># Generate density plots</span></span>
<span><span class="fu"><a href="reference/plotMarkerExpression.html">plotMarkerExpression</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">ref_data</span>, </span>
<span>                     query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                     reference_cell_labels <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                     query_cell_labels <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                     gene_name <span class="op">=</span> <span class="st">"MS4A1"</span>, </span>
<span>                     label <span class="op">=</span> <span class="st">"B_and_plasma"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Gene-Expression-Histogram-1.png" width="100%"></p>
<p>In the provided example, we examined the distribution of expression values for the gene MS4A1, a marker for naive B cells, in both the query and reference datasets. Additionally, we also looked at the distribution of MS4A1 expression in the B_and_plasma cell type. We observed overlapping distributions in both cases, suggesting alignment between the reference and query datasets.</p>
</div>
<div class="section level2">
<h2 id="evaluating-alignment-between-reference-and-query-datasets-in-terms-of-highly-variable-genes">Evaluating Alignment Between Reference and Query Datasets in Terms of Highly Variable Genes<a class="anchor" aria-label="anchor" href="#evaluating-alignment-between-reference-and-query-datasets-in-terms-of-highly-variable-genes"></a>
</h2>
<p>We are assessing the similarity or alignment between two datasets, the reference dataset, and the query dataset, in terms of highly variable genes (HVGs). We calculate the overlap coefficient between the sets of highly variable genes in the reference and query datasets. The overlap coefficient quantifies the degree of overlap or similarity between these two sets of genes. A value closer to 1 indicates a higher degree of overlap, while a value closer to 0 suggests less overlap. The computed overlap coefficient is printed, providing a numerical measure of how well the highly variable genes in the reference and query datasets align. In this case, the overlap coefficient is 0.63, indicating a moderate level of overlap.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Selecting highly variable genes</span></span>
<span><span class="va">ref_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">query_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the overlap coefficient</span></span>
<span><span class="va">overlap_coefficient</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calculateHVGOverlap.html">calculateHVGOverlap</a></span><span class="op">(</span>reference_genes <span class="op">=</span> <span class="va">ref_var</span>, </span>
<span>                                           query_genes <span class="op">=</span> <span class="va">query_var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">overlap_coefficient</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.63</span></span></code></pre></div>
<p>This analysis helps us assess the extent to which the reference and query datasets share highly variable genes, which can provide insights into the comparability and alignment of these datasets.</p>
</div>
<div class="section level2">
<h2 id="visualize-gene-expression-on-dimensional-reduction-plot">Visualize Gene Expression on Dimensional Reduction Plot<a class="anchor" aria-label="anchor" href="#visualize-gene-expression-on-dimensional-reduction-plot"></a>
</h2>
<p>To gain insights into the gene expression patterns and their representation in a dimensional reduction space, we can utilize the plotGeneExpressionDimred function. This function allows us to plot the gene expression values of a specific gene on a dimensional reduction plot generated using methods like t-SNE, UMAP, or PCA. Each single cell is color-coded based on its expression level of the gene of interest.</p>
<p>In the provided example, we are visualizing the gene expression values of the gene “VPREB3” on a PCA plot. The PCA plot represents the cells in a lower-dimensional space, where the x-axis corresponds to the first principal component (Dimension 1) and the y-axis corresponds to the second principal component (Dimension 2). Each cell is represented as a point on the plot, and its color reflects the expression level of the gene “VPREB3,” ranging from low (lighter color) to high (darker color).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate dimension reduction plot color code by gene expression</span></span>
<span><span class="fu"><a href="reference/plotGeneExpressionDimred.html">plotGeneExpressionDimred</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                         method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                         n_components <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, </span>
<span>                         feature <span class="op">=</span> <span class="st">"VPREB3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Gene-Expression-Scatter-1.png" width="100%"></p>
<p>The dimensional reduction plot allows us to observe how the gene expression of VPREB3 is distributed across the cells and whether any clusters or patterns emerge in the data.</p>
</div>
<div class="section level2">
<h2 id="visualize-gene-sets-or-pathway-scores-on-dimensional-reduction-plot">Visualize Gene Sets or Pathway Scores on Dimensional Reduction Plot<a class="anchor" aria-label="anchor" href="#visualize-gene-sets-or-pathway-scores-on-dimensional-reduction-plot"></a>
</h2>
<p>In addition to examining individual gene expression patterns, it is often useful to assess the collective activity of gene sets or pathways within single cells. This can provide insights into the functional states or biological processes associated with specific cell types or conditions. To facilitate this analysis, the scDiagnostics package includes a function called plotGeneSetScores that enables the visualization of gene set or pathway scores on a dimensional reduction plot.</p>
<p>The plotGeneSetScores function allows you to plot gene set or pathway scores on a dimensional reduction plot generated using methods such as PCA, t-SNE, or UMAP. Each single cell is color-coded based on its scores for specific gene sets or pathways. This visualization helps identify the heterogeneity and patterns of gene set or pathway activity within the dataset, potentially revealing subpopulations with distinct functional characteristics.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Compute scores using AUCell</span></span>
<span><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="st">"logcounts"</span><span class="op">)</span></span>
<span><span class="va">cells_rankings</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AUCell/man/AUCell_buildRankings.html" class="external-link">AUCell_buildRankings</a></span><span class="op">(</span><span class="va">expression_matrix</span>, plotStats <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate gene sets</span></span>
<span><span class="va">gene_set1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">gene_set2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>geneSet1 <span class="op">=</span> <span class="va">gene_set1</span>,</span>
<span>                  geneSet2 <span class="op">=</span> <span class="va">gene_set2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate AUC scores for gene sets</span></span>
<span><span class="va">cells_AUC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AUCell/man/AUCell_calcAUC.html" class="external-link">AUCell_calcAUC</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">cells_rankings</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">geneSetScores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">cells_AUC</span><span class="op">)</span><span class="op">[</span><span class="st">"geneSet1"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot gene set scores on PCA</span></span>
<span><span class="fu"><a href="reference/plotGeneSetScores.html">plotGeneSetScores</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                  feature <span class="op">=</span> <span class="st">"geneSetScores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Pathway-Scores-on-Dimensional-Reduction-Scatter-1.png" width="100%"></p>
<p>In the provided example, we demonstrate the usage of the plotGeneSetScores function using the AUCell package to compute gene set or pathway scores. Custom gene sets are generated for demonstration purposes, but users can provide their own gene set scores using any method of their choice. It is important to ensure that the scores are assigned to the colData of the reference or query object and specify the correct feature name for visualization.</p>
<p>By visualizing gene set or pathway scores on a dimensional reduction plot, you can gain a comprehensive understanding of the functional landscape within your single-cell gene expression dataset and explore the relationships between gene set activities and cellular phenotypes.</p>
</div>
<div class="section level2">
<h2 id="visualizing-reference-and-query-cell-types-using-multidimensional-scaling-mds">Visualizing Reference and Query Cell Types using Multidimensional Scaling (MDS)<a class="anchor" aria-label="anchor" href="#visualizing-reference-and-query-cell-types-using-multidimensional-scaling-mds"></a>
</h2>
<p>This function performs Multidimensional Scaling (MDS) analysis on the query and reference datasets to examine their similarity. The dissimilarity matrix is calculated based on the correlation between the datasets, representing the distances between cells in terms of gene expression patterns. MDS is then applied to derive low-dimensional coordinates for each cell. Subsequently, a scatter plot is generated, where each data point represents a cell, and cell types are color-coded using custom colors provided by the user. This visualization enables the comparison of cell type distributions between the query and reference datasets in a reduced-dimensional space.</p>
<p>The rationale behind this function is to visually assess the alignment and relationships between cell types in the query and reference datasets.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Intersect the gene symbols to obtain common genes</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_var</span>, <span class="va">query_var</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select desired cell types</span></span>
<span><span class="va">selected_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">ref_data_subset</span> <span class="op">&lt;-</span> <span class="va">ref_data</span><span class="op">[</span><span class="va">common_genes</span>, <span class="va">ref_data</span><span class="op">$</span><span class="va">reclustered.broad</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_cell_types</span><span class="op">]</span></span>
<span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span><span class="va">common_genes</span>, <span class="va">query_data</span><span class="op">$</span><span class="va">labels</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_cell_types</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Extract cell types for visualization</span></span>
<span><span class="va">ref_labels</span> <span class="op">&lt;-</span> <span class="va">ref_data_subset</span><span class="op">$</span><span class="va">reclustered.broad</span></span>
<span><span class="va">query_labels</span> <span class="op">&lt;-</span> <span class="va">query_data_subset</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Combine the cell type labels from both datasets</span></span>
<span><span class="va">mdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Query"</span>, <span class="va">query_labels</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Reference"</span>, <span class="va">ref_labels</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define the cell types and legend order</span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Query CD8"</span>, <span class="st">"Reference CD8"</span>, <span class="st">"Query CD4"</span>, <span class="st">"Reference CD4"</span>, <span class="st">"Query B_and_plasma"</span>, <span class="st">"Reference B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">legend_order</span> <span class="op">&lt;-</span> <span class="va">cell_types</span></span>
<span></span>
<span><span class="co">## Define the colors for cell types</span></span>
<span><span class="va">color_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_types</span><span class="op">)</span>, <span class="st">"Paired"</span><span class="op">)</span></span>
<span><span class="va">color_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">color_palette</span>, <span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">cell_type_colors</span> <span class="op">&lt;-</span> <span class="va">color_mapping</span><span class="op">[</span><span class="va">cell_types</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Generate the MDS scatter plot with cell type coloring</span></span>
<span><span class="fu"><a href="reference/visualizeCellTypeMDS.html">visualizeCellTypeMDS</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                     reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                     mdata <span class="op">=</span> <span class="va">mdata</span>, </span>
<span>                     cell_type_colors <span class="op">=</span> <span class="va">cell_type_colors</span>, </span>
<span>                     legend_order <span class="op">=</span> <span class="va">legend_order</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/CMD-Scatter-Plot-1.png" width="100%"></p>
<p>Upon examining the MDS scatter plot, we observe that the CD4 and CD8 cell types overlap to some extent.By observing the proximity or overlap of different cell types, one can gain insights into their potential relationships or shared characteristics.</p>
<p>The selection of custom genes and desired cell types depends on the user’s research interests and goals. It allows for flexibility in focusing on specific genes and examining particular cell types of interest in the visualization.</p>
</div>
<div class="section level2">
<h2 id="cell-type-specific-pairwise-correlation-analysis-and-visualization">Cell Type-specific Pairwise Correlation Analysis and Visualization<a class="anchor" aria-label="anchor" href="#cell-type-specific-pairwise-correlation-analysis-and-visualization"></a>
</h2>
<p>This analysis aims to explore the correlation patterns between different cell types in a single-cell gene expression dataset. The goal is to compare the gene expression profiles of cells from a reference dataset and a query dataset to understand the relationships and similarities between various cell types.</p>
<p>To perform the analysis, we start by computing the pairwise correlations between the query and reference cells for selected cell types (“CD4”, “CD8”, “B_and_plasma”). The Spearman correlation method is used, user can also use Pearsons correlation coeefficient.</p>
<p>This will return average correlation matrix which can be visulaized by user’s method of choice. Here, the results are visualized as a correlation plot using the corrplot package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu">computeAveragePairwiseCorrelation</span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                                    reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                                    query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                                    ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                                    cell_types <span class="op">=</span> <span class="va">selected_cell_types</span>, </span>
<span>                                                    correlation_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the pairwise average correlations using corrplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span>, method <span class="op">=</span> <span class="st">"number"</span>, tl.col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Cell-Type-Correlation-Analysis-Visualization-1.png" width="100%"></p>
<p>In this case, users have the flexibility to extract the gene expression profiles of specific cell types from the reference and query datasets and provide these profiles as input to the function. Additionally, they can select their own set of genes that they consider relevant for computing the pairwise correlations. For demonstartion we have used common highly variable genes from reference and query dataset.</p>
<p>By providing their own gene expression profiles and choosing specific genes, users can focus the analysis on the cell types and genes of interest to their research question.</p>
</div>
<div class="section level2">
<h2 id="pairwise-distance-analysis-and-density-visualization">Pairwise Distance Analysis and Density Visualization<a class="anchor" aria-label="anchor" href="#pairwise-distance-analysis-and-density-visualization"></a>
</h2>
<p>This function serves to conduct a analysis of pairwise distances or correlations between cells of specific cell types within a single-cell gene expression dataset. By calculating these distances or correlations, users can gain insights into the relationships and differences in gene expression profiles between different cell types. The function facilitates this analysis by generating density plots, allowing users to visualize the distribution of distances or correlations for various pairwise comparisons.</p>
<p>The analysis offers the flexibility to select a particular cell type for examination, and users can choose between different distance metrics, such as “euclidean” or “manhattan,” to calculate pairwise distances.</p>
<p>To illustrate, the function is applied to the cell type CD8 using the euclidean distance metric in the example below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/calculatePairwiseDistancesAndPlotDensity.html">calculatePairwiseDistancesAndPlotDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                         reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                         query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                         ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                         cell_type_query <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         cell_type_reference <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         distance_metric <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Pairwise-Distance-Analysis-Density-Visualization-1.png" width="100%"></p>
<p>Alternatively, users can opt for the “correlation” distance metric, which measures the similarity in gene expression profiles between cells.</p>
<p>To illustrate, the function is applied to the cell type CD8 using the correlation distance metric in the example below. By selecting either the “pearson” or “spearman” correlation method, users can emphasize either linear or rank-based associations, respectively.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/calculatePairwiseDistancesAndPlotDensity.html">calculatePairwiseDistancesAndPlotDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                         reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                         query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                         ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                         cell_type_query <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         cell_type_reference <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         distance_metric <span class="op">=</span> <span class="st">"correlation"</span>,</span>
<span>                                         correlation_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/Pairwise-Distance-Correlation-Based-Density-Visualization-1.png" width="100%"></p>
<p>By utilizing this function, users can explore the pairwise distances between query and reference cells of a specific cell type and gain insights into the distribution of distances through density plots. This analysis aids in understanding the similarities and differences in gene expression profiles for the selected cell type within the query and reference datasets.</p>
</div>
<div class="section level2">
<h2 id="linear-regression-analysis">Linear regression analysis<a class="anchor" aria-label="anchor" href="#linear-regression-analysis"></a>
</h2>
<p>Performing linear regression analysis on a SingleCellExperiment object enables users to examine the relationship between a principal component (PC) from the dimension reduction slot and an independent variable of interest. By specifying the desired dependent variable as one of the principal components (e.g., “PC1”, “PC2”, etc.) and providing the corresponding independent variable from the colData slot, users can explore the associations between these variables within the single-cell gene expression dataset (reference and query).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the dependent variables (principal components) and independent variable (e.g., "labels")</span></span>
<span><span class="va">dep.vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span><span class="op">)</span></span>
<span><span class="va">indep.var</span> <span class="op">&lt;-</span> <span class="st">"labels"</span></span>
<span></span>
<span><span class="co"># Perform linear regression on multiple principal components</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regressPC.html">regressPC</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">query_data</span>, dep.vars <span class="op">=</span> <span class="va">dep.vars</span>, indep.var <span class="op">=</span> <span class="va">indep.var</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Summaries of the linear regression models</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">regression_summaries</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co"># R-squared values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">rsquared_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co"># Variance contributions for each principal component</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">var_contributions_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co"># Total variance explained</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">total_variance_explained</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>By conducting linear regression, one can assess whether the PC values are significantly associated with the cell types. This analysis helps uncover whether there is a systematic variation in PC values across different cell types. Understanding the relationship between PC values and cell types can provide valuable insights into the biological or technical factors driving cellular heterogeneity. It can help identify PC dimensions that capture variation specific to certain cell types or distinguish different cellular states.</p>
</div>
</div>
<div class="section level1">
<h1 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h1>
<p>In this analysis, we have demonstrated the capabilities of the scDiagnostics package for assessing the appropriateness of cell assignments in single-cell gene expression profiles. By utilizing various diagnostic functions and visualization techniques, we have explored different aspects of the data, including total UMI counts, annotation scores, gene expression distributions, dimensional reduction plots, gene set scores, pairwise correlations, pairwise distances, and linear regression analysis.</p>
<p>Through the scatter plots, histograms, and dimensional reduction plots, we were able to gain insights into the relationships between gene expression patterns, cell types, and the distribution of cells in a reduced-dimensional space. The examination of gene expression distributions, gene sets, and pathways allowed us to explore the functional landscape and identify subpopulations with distinct characteristics within the dataset. Additionally, the pairwise correlation and distance analyses provided a deeper understanding of the similarities and differences between cell types, highlighting potential relationships and patterns.</p>
<hr>
<div class="section level2">
<h2 id="rsession-info">R.session Info<a class="anchor" aria-label="anchor" href="#rsession-info"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">3.1</span> (<span class="dv">2023-06-16</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>Platform<span class="sc">:</span> aarch64<span class="sc">-</span>apple<span class="sc">-</span><span class="fu">darwin20</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>Running under<span class="sc">:</span> macOS Sonoma <span class="fl">14.2</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.3</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRblas.<span class="fl">0.</span>dylib </span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.3</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.dylib;  LAPACK version <span class="dv">3</span>.<span class="fl">11.0</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>[<span class="dv">1</span>] en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>en_US.UTF<span class="dv">-8</span><span class="sc">/</span>en_US.UTF<span class="dv">-8</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>time zone<span class="sc">:</span> America<span class="sc">/</span>New_York</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a>tzcode source<span class="sc">:</span> internal</span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a>[<span class="dv">1</span>] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a>[<span class="dv">8</span>] base     </span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a> [<span class="dv">1</span>] celldex_1.<span class="fl">10.1</span>              corrplot_0<span class="fl">.92</span>              </span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a> [<span class="dv">3</span>] AUCell_1.<span class="fl">22.0</span>               SingleR_2.<span class="fl">2.0</span>              </span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a> [<span class="dv">5</span>] RColorBrewer_1<span class="fl">.1</span><span class="dv">-3</span>          scRNAseq_2.<span class="fl">14.0</span>            </span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a> [<span class="dv">7</span>] scran_1.<span class="fl">28.2</span>                scater_1.<span class="fl">29.2</span>              </span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a> [<span class="dv">9</span>] ggplot2_3.<span class="fl">4.4</span>               scuttle_1.<span class="fl">10.3</span>             </span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a>[<span class="dv">11</span>] scDiagnostics_0.<span class="fl">99.0</span>        SingleCellExperiment_1.<span class="fl">22.0</span></span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a>[<span class="dv">13</span>] SummarizedExperiment_1.<span class="fl">30.2</span> Biobase_2.<span class="fl">60.0</span>             </span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a>[<span class="dv">15</span>] GenomicRanges_1.<span class="fl">52.1</span>        GenomeInfoDb_1.<span class="fl">36.4</span>        </span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a>[<span class="dv">17</span>] IRanges_2.<span class="fl">34.1</span>              S4Vectors_0.<span class="fl">38.2</span>           </span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a>[<span class="dv">19</span>] BiocGenerics_0.<span class="fl">46.0</span>         MatrixGenerics_1.<span class="fl">12.3</span>      </span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a>[<span class="dv">21</span>] matrixStats_1.<span class="fl">0.0</span>          </span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a></span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb16-33"><a href="#cb16-33" tabindex="-1"></a>  [<span class="dv">1</span>] rstudioapi_0.<span class="fl">15.0</span>             magrittr_2.<span class="fl">0.3</span>               </span>
<span id="cb16-34"><a href="#cb16-34" tabindex="-1"></a>  [<span class="dv">3</span>] ggbeeswarm_0.<span class="fl">7.2</span>              GenomicFeatures_1.<span class="fl">52.2</span>       </span>
<span id="cb16-35"><a href="#cb16-35" tabindex="-1"></a>  [<span class="dv">5</span>] farver_2.<span class="fl">1.1</span>                  rmarkdown_2<span class="fl">.25</span>               </span>
<span id="cb16-36"><a href="#cb16-36" tabindex="-1"></a>  [<span class="dv">7</span>] BiocIO_1.<span class="fl">10.0</span>                 zlibbioc_1.<span class="fl">46.0</span>              </span>
<span id="cb16-37"><a href="#cb16-37" tabindex="-1"></a>  [<span class="dv">9</span>] vctrs_0.<span class="fl">6.4</span>                   Rsamtools_2.<span class="fl">16.0</span>             </span>
<span id="cb16-38"><a href="#cb16-38" tabindex="-1"></a> [<span class="dv">11</span>] memoise_2.<span class="fl">0.1</span>                 DelayedMatrixStats_1.<span class="fl">22.6</span>    </span>
<span id="cb16-39"><a href="#cb16-39" tabindex="-1"></a> [<span class="dv">13</span>] RCurl_1<span class="fl">.98-1.12</span>               htmltools_0.<span class="dv">5</span>.<span class="fl">8.1</span>            </span>
<span id="cb16-40"><a href="#cb16-40" tabindex="-1"></a> [<span class="dv">15</span>] S4Arrays_1.<span class="fl">0.6</span>                progress_1.<span class="fl">2.2</span>               </span>
<span id="cb16-41"><a href="#cb16-41" tabindex="-1"></a> [<span class="dv">17</span>] AnnotationHub_3.<span class="fl">8.0</span>           curl_5.<span class="fl">1.0</span>                   </span>
<span id="cb16-42"><a href="#cb16-42" tabindex="-1"></a> [<span class="dv">19</span>] BiocNeighbors_1.<span class="fl">18.0</span>          cachem_1.<span class="fl">0.8</span>                 </span>
<span id="cb16-43"><a href="#cb16-43" tabindex="-1"></a> [<span class="dv">21</span>] GenomicAlignments_1.<span class="fl">36.0</span>      igraph_1.<span class="fl">5.1</span>                 </span>
<span id="cb16-44"><a href="#cb16-44" tabindex="-1"></a> [<span class="dv">23</span>] mime_0<span class="fl">.12</span>                     lifecycle_1.<span class="fl">0.3</span>              </span>
<span id="cb16-45"><a href="#cb16-45" tabindex="-1"></a> [<span class="dv">25</span>] pkgconfig_2.<span class="fl">0.3</span>               rsvd_1.<span class="fl">0.5</span>                   </span>
<span id="cb16-46"><a href="#cb16-46" tabindex="-1"></a> [<span class="dv">27</span>] Matrix_1<span class="fl">.6-1.1</span>                R6_2.<span class="fl">5.1</span>                     </span>
<span id="cb16-47"><a href="#cb16-47" tabindex="-1"></a> [<span class="dv">29</span>] fastmap_1.<span class="fl">1.1</span>                 GenomeInfoDbData_1.<span class="fl">2.10</span>      </span>
<span id="cb16-48"><a href="#cb16-48" tabindex="-1"></a> [<span class="dv">31</span>] shiny_1.<span class="dv">7</span>.<span class="fl">5.1</span>                 digest_0.<span class="fl">6.33</span>                </span>
<span id="cb16-49"><a href="#cb16-49" tabindex="-1"></a> [<span class="dv">33</span>] colorspace_2<span class="fl">.1</span><span class="dv">-0</span>              AnnotationDbi_1.<span class="fl">62.2</span>         </span>
<span id="cb16-50"><a href="#cb16-50" tabindex="-1"></a> [<span class="dv">35</span>] dqrng_0.<span class="fl">3.1</span>                   irlba_2.<span class="dv">3</span>.<span class="fl">5.1</span>                </span>
<span id="cb16-51"><a href="#cb16-51" tabindex="-1"></a> [<span class="dv">37</span>] ExperimentHub_2.<span class="fl">8.1</span>           RSQLite_2.<span class="fl">3.1</span>                </span>
<span id="cb16-52"><a href="#cb16-52" tabindex="-1"></a> [<span class="dv">39</span>] beachmat_2.<span class="fl">16.0</span>               labeling_0.<span class="fl">4.3</span>               </span>
<span id="cb16-53"><a href="#cb16-53" tabindex="-1"></a> [<span class="dv">41</span>] filelock_1.<span class="fl">0.2</span>                fansi_1.<span class="fl">0.5</span>                  </span>
<span id="cb16-54"><a href="#cb16-54" tabindex="-1"></a> [<span class="dv">43</span>] httr_1.<span class="fl">4.7</span>                    abind_1<span class="fl">.4</span><span class="dv">-5</span>                  </span>
<span id="cb16-55"><a href="#cb16-55" tabindex="-1"></a> [<span class="dv">45</span>] compiler_4.<span class="fl">3.1</span>                bit64_4.<span class="fl">0.5</span>                  </span>
<span id="cb16-56"><a href="#cb16-56" tabindex="-1"></a> [<span class="dv">47</span>] withr_2.<span class="fl">5.1</span>                   BiocParallel_1.<span class="fl">34.2</span>          </span>
<span id="cb16-57"><a href="#cb16-57" tabindex="-1"></a> [<span class="dv">49</span>] viridis_0.<span class="fl">6.4</span>                 DBI_1.<span class="fl">1.3</span>                    </span>
<span id="cb16-58"><a href="#cb16-58" tabindex="-1"></a> [<span class="dv">51</span>] R.utils_2.<span class="fl">12.2</span>                biomaRt_2.<span class="fl">56.1</span>               </span>
<span id="cb16-59"><a href="#cb16-59" tabindex="-1"></a> [<span class="dv">53</span>] rappdirs_0.<span class="fl">3.3</span>                DelayedArray_0.<span class="fl">26.7</span>          </span>
<span id="cb16-60"><a href="#cb16-60" tabindex="-1"></a> [<span class="dv">55</span>] rjson_0.<span class="fl">2.21</span>                  bluster_1.<span class="fl">10.0</span>               </span>
<span id="cb16-61"><a href="#cb16-61" tabindex="-1"></a> [<span class="dv">57</span>] tools_4.<span class="fl">3.1</span>                   vipor_0.<span class="fl">4.5</span>                  </span>
<span id="cb16-62"><a href="#cb16-62" tabindex="-1"></a> [<span class="dv">59</span>] beeswarm_0.<span class="fl">4.0</span>                interactiveDisplayBase_1.<span class="fl">38.0</span></span>
<span id="cb16-63"><a href="#cb16-63" tabindex="-1"></a> [<span class="dv">61</span>] httpuv_1.<span class="fl">6.12</span>                 R.oo_1.<span class="fl">25.0</span>                  </span>
<span id="cb16-64"><a href="#cb16-64" tabindex="-1"></a> [<span class="dv">63</span>] glue_1.<span class="fl">6.2</span>                    restfulr_0.<span class="fl">0.15</span>              </span>
<span id="cb16-65"><a href="#cb16-65" tabindex="-1"></a> [<span class="dv">65</span>] promises_1.<span class="fl">2.1</span>                grid_4.<span class="fl">3.1</span>                   </span>
<span id="cb16-66"><a href="#cb16-66" tabindex="-1"></a> [<span class="dv">67</span>] cluster_2.<span class="fl">1.4</span>                 generics_0.<span class="fl">1.3</span>               </span>
<span id="cb16-67"><a href="#cb16-67" tabindex="-1"></a> [<span class="dv">69</span>] gtable_0.<span class="fl">3.4</span>                  R.methodsS3_1.<span class="fl">8.2</span>            </span>
<span id="cb16-68"><a href="#cb16-68" tabindex="-1"></a> [<span class="dv">71</span>] ensembldb_2.<span class="fl">24.1</span>              data.table_1.<span class="fl">14.8</span>            </span>
<span id="cb16-69"><a href="#cb16-69" tabindex="-1"></a> [<span class="dv">73</span>] hms_1.<span class="fl">1.3</span>                     xml2_1.<span class="fl">3.5</span>                   </span>
<span id="cb16-70"><a href="#cb16-70" tabindex="-1"></a> [<span class="dv">75</span>] BiocSingular_1.<span class="fl">16.0</span>           ScaledMatrix_1.<span class="fl">8.1</span>           </span>
<span id="cb16-71"><a href="#cb16-71" tabindex="-1"></a> [<span class="dv">77</span>] metapod_1.<span class="fl">8.0</span>                 utf8_1.<span class="fl">2.4</span>                   </span>
<span id="cb16-72"><a href="#cb16-72" tabindex="-1"></a> [<span class="dv">79</span>] XVector_0.<span class="fl">40.0</span>                ggrepel_0.<span class="fl">9.4</span>                </span>
<span id="cb16-73"><a href="#cb16-73" tabindex="-1"></a> [<span class="dv">81</span>] BiocVersion_3.<span class="fl">17.1</span>            pillar_1.<span class="fl">9.0</span>                 </span>
<span id="cb16-74"><a href="#cb16-74" tabindex="-1"></a> [<span class="dv">83</span>] stringr_1.<span class="fl">5.0</span>                 limma_3.<span class="fl">56.2</span>                 </span>
<span id="cb16-75"><a href="#cb16-75" tabindex="-1"></a> [<span class="dv">85</span>] later_1.<span class="fl">3.1</span>                   dplyr_1.<span class="fl">1.3</span>                  </span>
<span id="cb16-76"><a href="#cb16-76" tabindex="-1"></a> [<span class="dv">87</span>] BiocFileCache_2.<span class="fl">8.0</span>           lattice_0<span class="fl">.22</span><span class="dv">-5</span>               </span>
<span id="cb16-77"><a href="#cb16-77" tabindex="-1"></a> [<span class="dv">89</span>] rtracklayer_1.<span class="fl">60.1</span>            bit_4.<span class="fl">0.5</span>                    </span>
<span id="cb16-78"><a href="#cb16-78" tabindex="-1"></a> [<span class="dv">91</span>] annotate_1.<span class="fl">78.0</span>               tidyselect_1.<span class="fl">2.0</span>             </span>
<span id="cb16-79"><a href="#cb16-79" tabindex="-1"></a> [<span class="dv">93</span>] locfit_1<span class="fl">.5-9.8</span>                Biostrings_2.<span class="fl">68.1</span>            </span>
<span id="cb16-80"><a href="#cb16-80" tabindex="-1"></a> [<span class="dv">95</span>] knitr_1<span class="fl">.44</span>                    gridExtra_2<span class="fl">.3</span>                </span>
<span id="cb16-81"><a href="#cb16-81" tabindex="-1"></a> [<span class="dv">97</span>] ProtGenerics_1.<span class="fl">32.0</span>           edgeR_3.<span class="fl">42.4</span>                 </span>
<span id="cb16-82"><a href="#cb16-82" tabindex="-1"></a> [<span class="dv">99</span>] xfun_0<span class="fl">.40</span>                     statmod_1.<span class="fl">5.0</span>                </span>
<span id="cb16-83"><a href="#cb16-83" tabindex="-1"></a>[<span class="dv">101</span>] stringi_1.<span class="fl">7.12</span>                lazyeval_0.<span class="fl">2.2</span>               </span>
<span id="cb16-84"><a href="#cb16-84" tabindex="-1"></a>[<span class="dv">103</span>] yaml_2.<span class="fl">3.7</span>                    evaluate_0<span class="fl">.22</span>                </span>
<span id="cb16-85"><a href="#cb16-85" tabindex="-1"></a>[<span class="dv">105</span>] codetools_0<span class="fl">.2</span><span class="dv">-19</span>              tibble_3.<span class="fl">2.1</span>                 </span>
<span id="cb16-86"><a href="#cb16-86" tabindex="-1"></a>[<span class="dv">107</span>] graph_1.<span class="fl">78.0</span>                  BiocManager_1.<span class="fl">30.22</span>          </span>
<span id="cb16-87"><a href="#cb16-87" tabindex="-1"></a>[<span class="dv">109</span>] cli_3.<span class="fl">6.1</span>                     xtable_1<span class="fl">.8</span><span class="dv">-4</span>                 </span>
<span id="cb16-88"><a href="#cb16-88" tabindex="-1"></a>[<span class="dv">111</span>] munsell_0.<span class="fl">5.0</span>                 Rcpp_1.<span class="fl">0.11</span>                  </span>
<span id="cb16-89"><a href="#cb16-89" tabindex="-1"></a>[<span class="dv">113</span>] dbplyr_2.<span class="fl">3.4</span>                  png_0<span class="fl">.1</span><span class="dv">-8</span>                    </span>
<span id="cb16-90"><a href="#cb16-90" tabindex="-1"></a>[<span class="dv">115</span>] XML_3<span class="fl">.99-0.14</span>                 parallel_4.<span class="fl">3.1</span>               </span>
<span id="cb16-91"><a href="#cb16-91" tabindex="-1"></a>[<span class="dv">117</span>] ellipsis_0.<span class="fl">3.2</span>                blob_1.<span class="fl">2.4</span>                   </span>
<span id="cb16-92"><a href="#cb16-92" tabindex="-1"></a>[<span class="dv">119</span>] prettyunits_1.<span class="fl">2.0</span>             AnnotationFilter_1.<span class="fl">24.0</span>      </span>
<span id="cb16-93"><a href="#cb16-93" tabindex="-1"></a>[<span class="dv">121</span>] sparseMatrixStats_1.<span class="fl">12.2</span>      bitops_1<span class="fl">.0</span><span class="dv">-7</span>                 </span>
<span id="cb16-94"><a href="#cb16-94" tabindex="-1"></a>[<span class="dv">123</span>] GSEABase_1.<span class="fl">62.0</span>               viridisLite_0.<span class="fl">4.2</span>            </span>
<span id="cb16-95"><a href="#cb16-95" tabindex="-1"></a>[<span class="dv">125</span>] scales_1.<span class="fl">2.1</span>                  purrr_1.<span class="fl">0.2</span>                  </span>
<span id="cb16-96"><a href="#cb16-96" tabindex="-1"></a>[<span class="dv">127</span>] crayon_1.<span class="fl">5.2</span>                  rlang_1.<span class="fl">1.1</span>                  </span>
<span id="cb16-97"><a href="#cb16-97" tabindex="-1"></a>[<span class="dv">129</span>] cowplot_1.<span class="fl">1.1</span>                 KEGGREST_1.<span class="fl">40.1</span>              </span></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ccb-hms/scDiagnostics/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scDiagnostics</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Anthony Christidis <br><small class="roles"> Author, maintainer </small>  </li>
<li>Andrew Ghazi <br><small class="roles"> Author </small>  </li>
<li>Smriti Chawla <br><small class="roles"> Author </small>  </li>
<li>Ludwig Geistlinger <br><small class="roles"> Author </small>  </li>
<li>Robert Gentleman <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
