<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/scDiagnostics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="boxplotPCA.html">boxplotPCA()</a></code> </p>
        </td>
        <td><p>Plot Principal Components for Different Cell Types</p></td>
      </tr><tr><td>
          <p><code><a href="calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation()</a></code> <code><a href="calculateAveragePairwiseCorrelation.html">plot(<i>&lt;calculateAveragePairwiseCorrelation&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compute Average Pairwise Correlation between Cell Types</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCategorizationEntropy.html">calculateCategorizationEntropy()</a></code> </p>
        </td>
        <td><p>Calculate Categorization Entropy</p></td>
      </tr><tr><td>
          <p><code><a href="calculateDiscriminantSpace.html">calculateDiscriminantSpace()</a></code> <code><a href="calculateDiscriminantSpace.html">plot(<i>&lt;calculateDiscriminantSpace&gt;</i>)</a></code> </p>
        </td>
        <td><p>Project Query Data onto Discriminant Space of Reference Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateHVGOverlap.html">calculateHVGOverlap()</a></code> </p>
        </td>
        <td><p>Calculate the Overlap Coefficient for Highly Variable Genes</p></td>
      </tr><tr><td>
          <p><code><a href="calculateHotellingPValue.html">calculateHotellingPValue()</a></code> </p>
        </td>
        <td><p>Perform Hotelling's T-squared Test on PCA Scores for Single-cell RNA-seq Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateNearestNeighborProbabilities.html">calculateNearestNeighborProbabilities()</a></code> <code><a href="calculateNearestNeighborProbabilities.html">plot(<i>&lt;calculateNearestNeighborProbabilities&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculate Nearest Neighbor Diagnostics for Cell Type Classification</p></td>
      </tr><tr><td>
          <p><code><a href="calculateSampleDistances.html">calculateSampleDistances()</a></code> <code><a href="calculateSampleDistances.html">plot(<i>&lt;calculateSampleDistances&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compute Sample Distances Between Reference and Query Data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateSampleDistancesSimilarity.html">calculateSampleDistancesSimilarity()</a></code> </p>
        </td>
        <td><p>Function to Calculate Bhattacharyya Coefficients and Hellinger Distances</p></td>
      </tr><tr><td>
          <p><code><a href="calculateSampleSimilarityPCA.html">calculateSampleSimilarityPCA()</a></code> <code><a href="calculateSampleSimilarityPCA.html">plot(<i>&lt;calculateSampleSimilarityPCA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculate Sample Similarity Using PCA Loadings</p></td>
      </tr><tr><td>
          <p><code><a href="calculateVarImpOverlap.html">calculateVarImpOverlap()</a></code> </p>
        </td>
        <td><p>Compare Gene Importance Across Datasets Using Random Forest</p></td>
      </tr><tr><td>
          <p><code><a href="compareCCA.html">compareCCA()</a></code> <code><a href="compareCCA.html">plot(<i>&lt;compareCCA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Subspaces Spanned by Top Principal Components Using Canonical Correlation Analysis</p></td>
      </tr><tr><td>
          <p><code><a href="comparePCA.html">comparePCA()</a></code> <code><a href="comparePCA.html">plot(<i>&lt;comparePCA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Principal Components Analysis (PCA) Results</p></td>
      </tr><tr><td>
          <p><code><a href="comparePCASubspace.html">comparePCASubspace()</a></code> <code><a href="comparePCASubspace.html">plot(<i>&lt;comparePCASubspace&gt;</i>)</a></code> </p>
        </td>
        <td><p>Compare Subspaces Spanned by Top Principal Components</p></td>
      </tr><tr><td>
          <p><code><a href="detectAnomaly.html">detectAnomaly()</a></code> <code><a href="detectAnomaly.html">plot(<i>&lt;detectAnomaly&gt;</i>)</a></code> </p>
        </td>
        <td><p>PCA Anomaly Scores via Isolation Forests with Visualization</p></td>
      </tr><tr><td>
          <p><code><a href="histQCvsAnnotation.html">histQCvsAnnotation()</a></code> </p>
        </td>
        <td><p>Histograms: QC Stats and Annotation Scores Visualization</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneExpressionDimred.html">plotGeneExpressionDimred()</a></code> </p>
        </td>
        <td><p>Visualize gene expression on a dimensional reduction plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneSetScores.html">plotGeneSetScores()</a></code> </p>
        </td>
        <td><p>Visualization of gene sets or pathway scores on dimensional reduction plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotMarkerExpression.html">plotMarkerExpression()</a></code> </p>
        </td>
        <td><p>Plot gene expression distribution from overall and cell type-specific perspective</p></td>
      </tr><tr><td>
          <p><code><a href="plotPairwiseDistancesDensity.html">plotPairwiseDistancesDensity()</a></code> </p>
        </td>
        <td><p>Pairwise Distance Analysis and Density Visualization</p></td>
      </tr><tr><td>
          <p><code><a href="plotQCvsAnnotation.html">plotQCvsAnnotation()</a></code> </p>
        </td>
        <td><p>Scatter plot: QC stats vs Cell Type Annotation Scores</p></td>
      </tr><tr><td>
          <p><code><a href="projectPCA.html">projectPCA()</a></code> </p>
        </td>
        <td><p>Project Query Data Onto PCA Space of Reference Data</p></td>
      </tr><tr><td>
          <p><code><a href="regressPC.html">regressPC()</a></code> <code><a href="regressPC.html">plotPCRegression()</a></code> </p>
        </td>
        <td><p>Principal component regression</p></td>
      </tr><tr><td>
          <p><code><a href="visualizeCellTypeMDS.html">visualizeCellTypeMDS()</a></code> </p>
        </td>
        <td><p>Visualizing Reference and Query Cell Types using MDS</p></td>
      </tr><tr><td>
          <p><code><a href="visualizeCellTypePCA.html">visualizeCellTypePCA()</a></code> </p>
        </td>
        <td><p>Visualize Principal Components for Different Cell Types</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

