<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Principal component regression — regressPC • scDiagnostics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Principal component regression — regressPC"><meta property="og:description" content="This function performs linear regression of a covariate of interest onto one
or more principal components, based on the data in a SingleCellExperiment
object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/scDiagnostics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Principal component regression</h1>
    <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/HEAD/R/regressPC.R" class="external-link"><code>R/regressPC.R</code></a></small>
    <div class="hidden name"><code>regressPC.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs linear regression of a covariate of interest onto one
or more principal components, based on the data in a SingleCellExperiment
object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regressPC</span><span class="op">(</span><span class="va">sce</span>, dep.vars <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">indep.var</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotPCRegression</span><span class="op">(</span><span class="va">sce</span>, <span class="va">regressPC_res</span>, dep.vars <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">indep.var</span>, max_pc <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sce</dt>
<dd><p>An object of class <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code>
containing the data for regression analysis.</p></dd>


<dt>dep.vars</dt>
<dd><p>character. Dependent variable(s). Determines which principal
component(s) (e.g., "PC1", "PC2", etc.) are used as explanatory variables.
Principal components are expected to be stored in a PC matrix named
<code>"PCA"</code> in the <code>reducedDims</code> of <code>sce</code>. Defaults to
<code>NULL</code> which will then regress on each principal component present in
the PC matrix.</p></dd>


<dt>indep.var</dt>
<dd><p>A character with the name of the independent variable. A column name in the
<code>colData</code> of <code>sce</code> specifying the response variable.</p></dd>


<dt>regressPC_res</dt>
<dd><p>a result from <code>regressPC</code></p></dd>


<dt>max_pc</dt>
<dd><p>The maximum number of PCs to show on the plot. Set to 0 to show all.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>list</code> containing</p><ul><li><p>summaries of the linear
  regression models for each specified principal component,</p></li>
<li><p>the
  corresponding R-squared (R2) values,</p></li>
<li><p>the variance contributions for
  each principal component, and</p></li>
<li><p>the total variance explained.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Principal component regression, derived from PCA, can be used to quantify the 
variance explained by a covariate of interest. Applications for single-cell 
analysis include quantification of batch effects, assessing clustering 
homogeneity, and evaluating alignment of query and reference datasets in cell 
type annotation settings.</p>
<p>Briefly, the \(R^2\) is calculated from a linear regression of the covariate B of 
interest onto each principal component. The variance contribution of the 
covariate effect per principal component is then calculated as the product of 
the variance explained by the i-th principal component (PC) and the 
corresponding \(R^2(PC_i | B)\). The sum across all variance contributions by the 
covariate effects in all principal components gives the total variance 
explained by the covariate as follows:</p>
<p>$$Var(C|B) = \sum_{i=1}^G \text{Var}(C|PC_i) \times R^2(PC_i | B)$$</p>
<p>where, \(\text{Var}(C \mid PC_i)\) is the variance of the data matrix \(C\) 
explained by the i-th principal component. See references for details.</p>
<p>If the input is large (&gt;3e4 cells) and the independent variable is categorical 
with \(&gt;10\) categories, this function will use a stripped-down linear model 
function that is faster but doesn't return all the same components. Namely, 
the <code>regression.summaries</code> component of the result will contain only the 
\(R^2\) values, nothing else.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Luecken et al. Benchmarking atlas-level data integration in
  single-cell genomics. Nature Methods, 19:41-50, 2022.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load data</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HeOrganAtlasData.html" class="external-link">HeOrganAtlasData</a></span><span class="op">(</span>tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Marrow"</span><span class="op">)</span>, ensembl <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Divide the data into reference and query datasets</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.7</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    replace <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">indices</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">-</span><span class="va">indices</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># log transform datasets</span></span></span>
<span class="r-in"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ref</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run PCA</span></span></span>
<span class="r-in"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get cell type scores using SingleR</span></span></span>
<span class="r-in"><span><span class="co"># Note: replace when using cell type annotation scores from other methods</span></span></span>
<span class="r-in"><span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">ref</span>, labels <span class="op">=</span> <span class="va">ref</span><span class="op">$</span><span class="va">reclustered.broad</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Add labels to query object</span></span></span>
<span class="r-in"><span><span class="va">query</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">scores</span><span class="op">$</span><span class="va">labels</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Specify the dependent variables (principal components) and</span></span></span>
<span class="r-in"><span><span class="co"># independent variable (e.g., "labels")</span></span></span>
<span class="r-in"><span><span class="va">dep.vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">indep.var</span> <span class="op">&lt;-</span> <span class="st">"labels"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform linear regression on multiple principal components</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">regressPC</span><span class="op">(</span></span></span>
<span class="r-in"><span>    sce <span class="op">=</span> <span class="va">query</span>,</span></span>
<span class="r-in"><span>    dep.vars <span class="op">=</span> <span class="va">dep.vars</span>,</span></span>
<span class="r-in"><span>    indep.var <span class="op">=</span> <span class="va">indep.var</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Obtain linear regression summaries and R-squared values</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">regression.summaries</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $PC1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lm(formula = f, data = df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residuals:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Min       1Q   Median       3Q      Max </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -14.4118  -2.5919  -0.3286   2.5929  12.2589 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Coefficients:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate Std. Error t value Pr(&gt;|t|)    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (Intercept)         -9.6062     0.3500  -27.45   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD4       7.7054     0.4108   18.76   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD8      15.3986     0.4271   36.05   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentMyeloid  11.9732     0.4753   25.19   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ---</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residual standard error: 4.156 on 965 degrees of freedom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Multiple R-squared:  0.5996,	Adjusted R-squared:  0.5984 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> F-statistic: 481.7 on 3 and 965 DF,  p-value: &lt; 2.2e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $PC2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lm(formula = f, data = df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residuals:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Min       1Q   Median       3Q      Max </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -10.1218  -1.5210  -0.3487   0.9409  16.9802 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Coefficients:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate Std. Error t value Pr(&gt;|t|)    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (Intercept)          4.2411     0.2548   16.64   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD4      -7.5991     0.2991  -25.40   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD8      -4.6103     0.3110  -14.82   &lt;2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentMyeloid   0.3151     0.3461    0.91    0.363    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ---</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residual standard error: 3.026 on 965 degrees of freedom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Multiple R-squared:  0.537,	Adjusted R-squared:  0.5356 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> F-statistic: 373.1 on 3 and 965 DF,  p-value: &lt; 2.2e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $PC3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> lm(formula = f, data = df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residuals:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Min      1Q  Median      3Q     Max </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -8.0409 -1.6677 -0.0591  1.6435  9.4789 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Coefficients:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Estimate Std. Error t value Pr(&gt;|t|)    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (Intercept)          2.8954     0.2138   13.54  &lt; 2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD4      -4.2704     0.2509  -17.02  &lt; 2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentCD8      -0.7463     0.2609   -2.86  0.00432 ** </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IndependentMyeloid  -5.9754     0.2903  -20.58  &lt; 2e-16 ***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ---</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Residual standard error: 2.538 on 965 degrees of freedom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Multiple R-squared:  0.4357,	Adjusted R-squared:  0.434 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> F-statistic: 248.4 on 3 and 965 DF,  p-value: &lt; 2.2e-16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">rsquared</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       PC1       PC2       PC3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.5996141 0.5370214 0.4357447 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">plotPCRegression</span><span class="op">(</span><span class="va">query</span>, <span class="va">res</span>, <span class="va">dep.vars</span>, <span class="va">indep.var</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 17 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_point()`).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 17 rows containing missing values or values outside the scale range</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> (`geom_line()`).</span>
<span class="r-plt img"><img src="regressPC-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

