<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scDiagnostics: diagnostic functions to assess the quality of cell type assignments in single-cell RNA-seq • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="scDiagnostics: diagnostic functions to assess the quality of cell type assignments in single-cell RNA-seq">
<meta property="og:description" content="scDiagnostics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scDiagnostics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scDiagnostics: diagnostic functions to assess
the quality of cell type assignments in single-cell RNA-seq</h1>
                        <h4 data-toc-skip class="author">Smriti
Chawla</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/HEAD/vignettes/scDiagnostics.Rmd" class="external-link"><code>vignettes/scDiagnostics.Rmd</code></a></small>
      <div class="hidden name"><code>scDiagnostics.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install the package from Bioconductor use a following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>To install the development version from Github:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: you will need the remotes package to install from github.</p>
<p>To build the package vignettes upon installation use:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span>,</span>
<span>                     build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p>R version tested: 4.2.3 (2023-03-15) or above</p>
<p>To use the scDiagnostics package, you need the following R packages
installed:</p>
<ul>
<li>SingleCellExperiment</li>
<li>scRNAseq</li>
<li>scater</li>
<li>scran</li>
<li>ggplot2</li>
<li>AUCell</li>
<li>SingleR</li>
<li>Matrix</li>
<li>corrplot</li>
<li>celldex</li>
<li>RColorBrewer</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>To explore the capabilities of the scDiagnostics package, you can
load your own data or utilize publicly available datasets obtained from
the scRNAseq R package. In this guide, we will demonstrate how to use
scDiagnostics with such datasets, which serve as valuable resources for
exploring the package and assessing the appropriateness of cell type
assignments.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Loading libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://scenic.aertslab.org" class="external-link">AUCell</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex" class="external-link">celldex</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="scatter-plot-qc-stats-vs--annotation-scores">Scatter Plot: QC stats vs. Annotation Scores<a class="anchor" aria-label="anchor" href="#scatter-plot-qc-stats-vs--annotation-scores"></a>
</h3>
<p>Here, we will consider the Human Primary Cell Atlas (Mabbott et
al. 2013) as a reference dataset and our query dataset consists of
Haematopoietic stem and progenitor cells from (Bunis DG et
al. 2021).</p>
<p>In scRNA-seq studies, assessing the quality of cells is important for
accurate downstream analyses. At the same time, assigning accurate cell
type labels based on gene expression profiles is an integral aspect of
scRNA-seq data interpretation. Generally, these two are performed
independently of each other. The rationale behind this function is to
inspect whether certain QC (Quality Control) criteria impact the
confidence level of cell type annotations.</p>
<p>For instance, it is reasonable to hypothesize that higher library
sizes could contribute to increased annotation confidence due to
enhanced statistical power for identifying cell type-specific gene
expression patterns, as evident in the scatter plot below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># load reference dataset</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/HumanPrimaryCellAtlasData.html" class="external-link">HumanPrimaryCellAtlasData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load query dataset (Bunis haematopoietic stem and progenitor cell</span></span>
<span><span class="co"># data) from Bunis DG et al. (2021). Single-Cell Mapping of</span></span>
<span><span class="co"># Progressive Fetal-to-Adult Transition in Human Naive T Cells Cell</span></span>
<span><span class="co"># Rep. 34(1): 108573</span></span>
<span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/BunisHSPCData.html" class="external-link">BunisHSPCData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span></span>
<span></span>
<span><span class="co"># Add QC metrics to query data</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Log transform query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run SingleR to predict cell types</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign predicted labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Get annotation scores</span></span>
<span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">scores</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">cell_scores</span> <span class="op">&lt;-</span> <span class="va">scores</span></span>
<span></span>
<span><span class="co"># Create a scatter plot between library size and annotation scores</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    qc_col <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>    label_col <span class="op">=</span> <span class="st">"pred.labels"</span>,</span>
<span>    score_col <span class="op">=</span> <span class="st">"cell_scores"</span>,</span>
<span>    label <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Scatter-Plot-LibrarySize-Vs-Annotation-Scores-1.png" width="700"></p>
<p>However, certain QC metrics, such as the proportion of mitochondrial
genes, may require careful consideration as they can sometimes be
associated with cellular states or functions rather than noise. The
interpretation of mitochondrial content should be context-specific and
informed by biological knowledge.</p>
<p>In next analysis, we investigated the relationship between
mitochondrial percentage and cell type annotation scores using liver
tissue data from He S et al. 2020. Notably, we observed high annotation
scores for macrophages and monocytes. These findings align with the
established biological characteristic of high mitochondrial activity in
macrophages and monocytes, adding biological context to our results.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HeOrganAtlasData.html" class="external-link">HeOrganAtlasData</a></span><span class="op">(</span></span>
<span>    tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Liver"</span><span class="op">)</span>,</span>
<span>    ensembl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    location <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add QC metrics to query data</span></span>
<span></span>
<span><span class="va">mito_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">unfiltered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">query_data</span>,subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mt <span class="op">=</span> <span class="va">mito_genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/quickPerCellQC.html" class="external-link">quickPerCellQC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span>, sub.fields <span class="op">=</span> <span class="st">"subsets_mt_percent"</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>,<span class="op">!</span><span class="va">qc</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Log transform query dataset</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run SingleR to predict cell types</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign predicted labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Get annotation scores</span></span>
<span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">scores</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">cell_scores</span> <span class="op">&lt;-</span> <span class="va">scores</span></span>
<span></span>
<span><span class="co"># Create a new column for the labels so it is easy to distinguish</span></span>
<span><span class="co">#  between Macrophoges, Monocytes and other cells</span></span>
<span><span class="va">query_data</span><span class="op">$</span><span class="va">label_category</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">$</span><span class="va">pred.labels</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>, <span class="st">"Monocyte"</span><span class="op">)</span>,</span>
<span>           <span class="va">query_data</span><span class="op">$</span><span class="va">pred.labels</span>,</span>
<span>           <span class="st">"Other cells"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Define custom colors for cell type labels</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Other cells"</span> <span class="op">=</span> <span class="st">"grey"</span>, <span class="st">"Macrophage"</span> <span class="op">=</span> <span class="st">"green"</span>, <span class="st">"Monocyte"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate scatter plot for all cell types</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotQCvsAnnotation.html">plotQCvsAnnotation</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>    qc_col <span class="op">=</span> <span class="st">"subsets_mt_percent"</span>,</span>
<span>    label_col <span class="op">=</span> <span class="st">"label_category"</span>,</span>
<span>    score_col <span class="op">=</span> <span class="st">"cell_scores"</span>,</span>
<span>    label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"subsets_mt_percent"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/QC-Annotation-Scatter-Mito-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="examining-distribution-of-qc-stats-and-annotation-scores">Examining Distribution of QC stats and Annotation Scores<a class="anchor" aria-label="anchor" href="#examining-distribution-of-qc-stats-and-annotation-scores"></a>
</h3>
<p>In addition to the scatter plot, we can gain further insights into
the gene expression profiles by visualizing the distribution of user
defined QC stats and annotation scores for all the cell types or
specific cell types. This allows us to examine the variation and
patterns in expression levels and scores across cells assigned to the
cell type of interest.</p>
<p>To accomplish this, we create two separate histograms. The first
histogram displays the distribution of the annotation scores.</p>
<p>The second histogram visualizes the distribution of QC stats. This
provides insights into the overall gene expression levels for the
specific cell type. Here in this particular example we are investigating
percentage of mitochondrial genes.</p>
<p>By examining the histograms, we can observe the range, shape, and
potential outliers in the distribution of both annotation scores and QC
stats. This allows us to assess the appropriateness of the cell type
assignments and identify any potential discrepancies or patterns in the
gene expression profiles for the specific cell type.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate histogram</span></span>
<span><span class="fu"><a href="../reference/histQCvsAnnotation.html">histQCvsAnnotation</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, qc_col <span class="op">=</span> <span class="st">"subsets_mt_percent"</span>, </span>
<span>                   label_col <span class="op">=</span> <span class="st">"pred.labels"</span>, </span>
<span>                   score_col <span class="op">=</span> <span class="st">"cell_scores"</span>, </span>
<span>                   label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Mito-Genes-Vs-Annotation-1.png" width="700"></p>
<p>The right-skewed distribution for mitochondrial percentages and a
left-skewed distribution for annotation scores in above histograms
suggest that most cells have lower mitochondrial contamination and
higher confidence in their assigned cell types.</p>
</div>
<div class="section level3">
<h3 id="exploring-gene-expression-distribution">Exploring Gene Expression Distribution<a class="anchor" aria-label="anchor" href="#exploring-gene-expression-distribution"></a>
</h3>
<p>This function helps user to explore the distribution of gene
expression values for a specific gene of interest across all the cells
in both reference and query datasets and within specific cell types.
This helps to evaluate whether the distributions are similar or aligned
between the datasets. Discrepancies in distribution patterns may
indicate potential incompatibilities or differences between the
datasets.</p>
<p>The function also allows users to narrow down their analysis to
specific cell types of interest. This enables investigation of whether
alignment between the query and reference datasets is consistent not
only at a global level but also within specific cell types.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Load data</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/HeOrganAtlasData.html" class="external-link">HeOrganAtlasData</a></span><span class="op">(</span>tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Marrow"</span><span class="op">)</span>, ensembl <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Divide the data into reference and query datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.7</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">indices</span><span class="op">]</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">-</span><span class="va">indices</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Log-transform datasets</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ref_data</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run PCA</span></span>
<span><span class="va">ref_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data</span><span class="op">)</span></span>
<span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get cell type scores using SingleR</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="va">ref_data</span>, labels <span class="op">=</span> <span class="va">ref_data</span><span class="op">$</span><span class="va">reclustered.broad</span><span class="op">)</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pred</span><span class="op">)</span></span>
<span>   </span>
<span><span class="co"># Assign labels to query data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">pred</span><span class="op">$</span><span class="va">labels</span></span>
<span>   </span>
<span><span class="co"># Generate density plots</span></span>
<span><span class="fu"><a href="../reference/plotMarkerExpression.html">plotMarkerExpression</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">ref_data</span>, </span>
<span>                     query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                     reference_cell_labels <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                     query_cell_labels <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                     gene_name <span class="op">=</span> <span class="st">"MS4A1"</span>, </span>
<span>                     label <span class="op">=</span> <span class="st">"B_and_plasma"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Gene-Expression-Histogram-1.png" width="700"></p>
<p>In the provided example, we examined the distribution of expression
values for the gene MS4A1, a marker for naive B cells, in both the query
and reference datasets. Additionally, we also looked at the distribution
of MS4A1 expression in the B_and_plasma cell type. We observed
overlapping distributions in both cases, suggesting alignment between
the reference and query datasets.</p>
</div>
<div class="section level3">
<h3 id="evaluating-alignment-between-reference-and-query-datasets-in-terms-of-highly-variable-genes">Evaluating Alignment Between Reference and Query Datasets in Terms
of Highly Variable Genes<a class="anchor" aria-label="anchor" href="#evaluating-alignment-between-reference-and-query-datasets-in-terms-of-highly-variable-genes"></a>
</h3>
<p>We are assessing the similarity or alignment between two datasets,
the reference dataset, and the query dataset, in terms of highly
variable genes (HVGs). We calculate the overlap coefficient between the
sets of highly variable genes in the reference and query datasets. The
overlap coefficient quantifies the degree of overlap or similarity
between these two sets of genes. A value closer to 1 indicates a higher
degree of overlap, while a value closer to 0 suggests less overlap. The
computed overlap coefficient is printed, providing a numerical measure
of how well the highly variable genes in the reference and query
datasets align. In this case, the overlap coefficient is 0.62,
indicating a moderate level of overlap.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Selecting highly variable genes</span></span>
<span><span class="va">ref_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">query_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data</span>, n<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the overlap coefficient</span></span>
<span><span class="va">overlap_coefficient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVGOverlap.html">calculateHVGOverlap</a></span><span class="op">(</span>reference_genes <span class="op">=</span> <span class="va">ref_var</span>, </span>
<span>                                           query_genes <span class="op">=</span> <span class="va">query_var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">overlap_coefficient</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.63</span></span></code></pre></div>
<p>In the provided example, we examined the distribution of expression
values for the gene MS4A1, a marker for naive B cells, in both the query
and reference datasets. Additionally, we also looked at the distribution
of MS4A1 expression in the B_and_plasma cell type. We observed
overlapping distributions in both cases, suggesting alignment between
the reference and query datasets.</p>
</div>
<div class="section level3">
<h3 id="visualize-gene-expression-on-dimensional-reduction-plot">Visualize Gene Expression on Dimensional Reduction Plot<a class="anchor" aria-label="anchor" href="#visualize-gene-expression-on-dimensional-reduction-plot"></a>
</h3>
<p>To gain insights into the gene expression patterns and their
representation in a dimensional reduction space, we can utilize the
plotGeneExpressionDimred function. This function allows us to plot the
gene expression values of a specific gene on a dimensional reduction
plot generated using methods like t-SNE, UMAP, or PCA. Each single cell
is color-coded based on its expression level of the gene of
interest.</p>
<p>In the provided example, we are visualizing the gene expression
values of the gene “VPREB3” on a PCA plot. The PCA plot represents the
cells in a lower-dimensional space, where the x-axis corresponds to the
first principal component (Dimension 1) and the y-axis corresponds to
the second principal component (Dimension 2). Each cell is represented
as a point on the plot, and its color reflects the expression level of
the gene “VPREB3,” ranging from low (lighter color) to high (darker
color).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate dimension reduction plot color code by gene expression</span></span>
<span><span class="fu"><a href="../reference/plotGeneExpressionDimred.html">plotGeneExpressionDimred</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                         method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                         n_components <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, </span>
<span>                         feature <span class="op">=</span> <span class="st">"VPREB3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Gene-Expression-Scatter-1.png" width="700"></p>
<p>The dimensional reduction plot allows us to observe how the gene
expression of VPREB3 is distributed across the cells and whether any
clusters or patterns emerge in the data.</p>
</div>
<div class="section level3">
<h3 id="visualize-gene-sets-or-pathway-scores-on-dimensional-reduction-plot">Visualize Gene Sets or Pathway Scores on Dimensional Reduction
Plot<a class="anchor" aria-label="anchor" href="#visualize-gene-sets-or-pathway-scores-on-dimensional-reduction-plot"></a>
</h3>
<p>In addition to examining individual gene expression patterns, it is
often useful to assess the collective activity of gene sets or pathways
within single cells. This can provide insights into the functional
states or biological processes associated with specific cell types or
conditions. To facilitate this analysis, the scDiagnostics package
includes a function called plotGeneSetScores that enables the
visualization of gene set or pathway scores on a dimensional reduction
plot.</p>
<p>The plotGeneSetScores function allows you to plot gene set or pathway
scores on a dimensional reduction plot generated using methods such as
PCA, t-SNE, or UMAP. Each single cell is color-coded based on its scores
for specific gene sets or pathways. This visualization helps identify
the heterogeneity and patterns of gene set or pathway activity within
the dataset, potentially revealing subpopulations with distinct
functional characteristics.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Compute scores using AUCell</span></span>
<span><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">query_data</span>, <span class="st">"logcounts"</span><span class="op">)</span></span>
<span><span class="va">cells_rankings</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AUCell/man/AUCell_buildRankings.html" class="external-link">AUCell_buildRankings</a></span><span class="op">(</span><span class="va">expression_matrix</span>, plotStats <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate gene sets</span></span>
<span><span class="va">gene_set1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">gene_set2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>geneSet1 <span class="op">=</span> <span class="va">gene_set1</span>,</span>
<span>                  geneSet2 <span class="op">=</span> <span class="va">gene_set2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate AUC scores for gene sets</span></span>
<span><span class="va">cells_AUC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AUCell/man/AUCell_calcAUC.html" class="external-link">AUCell_calcAUC</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">cells_rankings</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign scores to colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">geneSetScores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">cells_AUC</span><span class="op">)</span><span class="op">[</span><span class="st">"geneSet1"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot gene set scores on PCA</span></span>
<span><span class="fu"><a href="../reference/plotGeneSetScores.html">plotGeneSetScores</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                  feature <span class="op">=</span> <span class="st">"geneSetScores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Pathway-Scores-on-Dimensional-Reduction-Scatter-1.png" width="700"></p>
<p>In the provided example, we demonstrate the usage of the
plotGeneSetScores function using the AUCell package to compute gene set
or pathway scores. Custom gene sets are generated for demonstration
purposes, but users can provide their own gene set scores using any
method of their choice. It is important to ensure that the scores are
assigned to the colData of the reference or query object and specify the
correct feature name for visualization.</p>
<p>By visualizing gene set or pathway scores on a dimensional reduction
plot, you can gain a comprehensive understanding of the functional
landscape within your single-cell gene expression dataset and explore
the relationships between gene set activities and cellular
phenotypes.</p>
</div>
<div class="section level3">
<h3 id="visualizing-reference-and-query-cell-types-using-multidimensional-scaling-mds">Visualizing Reference and Query Cell Types using Multidimensional
Scaling (MDS)<a class="anchor" aria-label="anchor" href="#visualizing-reference-and-query-cell-types-using-multidimensional-scaling-mds"></a>
</h3>
<p>This function performs Multidimensional Scaling (MDS) analysis on the
query and reference datasets to examine their similarity. The
dissimilarity matrix is calculated based on the correlation between the
datasets, representing the distances between cells in terms of gene
expression patterns. MDS is then applied to derive low-dimensional
coordinates for each cell. Subsequently, a scatter plot is generated,
where each data point represents a cell, and cell types are color-coded
using custom colors provided by the user. This visualization enables the
comparison of cell type distributions between the query and reference
datasets in a reduced-dimensional space.</p>
<p>The rationale behind this function is to visually assess the
alignment and relationships between cell types in the query and
reference datasets.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Intersect the gene symbols to obtain common genes</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_var</span>, <span class="va">query_var</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select desired cell types</span></span>
<span><span class="va">selected_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">ref_data_subset</span> <span class="op">&lt;-</span> <span class="va">ref_data</span><span class="op">[</span><span class="va">common_genes</span>, <span class="va">ref_data</span><span class="op">$</span><span class="va">reclustered.broad</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_cell_types</span><span class="op">]</span></span>
<span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span><span class="va">common_genes</span>, <span class="va">query_data</span><span class="op">$</span><span class="va">labels</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_cell_types</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Extract cell types for visualization</span></span>
<span><span class="va">ref_labels</span> <span class="op">&lt;-</span> <span class="va">ref_data_subset</span><span class="op">$</span><span class="va">reclustered.broad</span></span>
<span><span class="va">query_labels</span> <span class="op">&lt;-</span> <span class="va">query_data_subset</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Combine the cell type labels from both datasets</span></span>
<span><span class="va">mdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Query"</span>, <span class="va">query_labels</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Reference"</span>, <span class="va">ref_labels</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define the cell types and legend order</span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Query CD8"</span>, <span class="st">"Reference CD8"</span>, <span class="st">"Query CD4"</span>, <span class="st">"Reference CD4"</span>, <span class="st">"Query B_and_plasma"</span>, <span class="st">"Reference B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">legend_order</span> <span class="op">&lt;-</span> <span class="va">cell_types</span></span>
<span></span>
<span><span class="co">## Define the colors for cell types</span></span>
<span><span class="va">color_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_types</span><span class="op">)</span>, <span class="st">"Paired"</span><span class="op">)</span></span>
<span><span class="va">color_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">color_palette</span>, <span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">cell_type_colors</span> <span class="op">&lt;-</span> <span class="va">color_mapping</span><span class="op">[</span><span class="va">cell_types</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Generate the MDS scatter plot with cell type coloring</span></span>
<span><span class="fu"><a href="../reference/visualizeCellTypeMDS.html">visualizeCellTypeMDS</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                     reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                     mdata <span class="op">=</span> <span class="va">mdata</span>, </span>
<span>                     cell_type_colors <span class="op">=</span> <span class="va">cell_type_colors</span>, </span>
<span>                     legend_order <span class="op">=</span> <span class="va">legend_order</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/CMD-Scatter-Plot-1.png" width="700"></p>
<p>Upon examining the MDS scatter plot, we observe that the CD4 and CD8
cell types overlap to some extent.By observing the proximity or overlap
of different cell types, one can gain insights into their potential
relationships or shared characteristics.</p>
<p>The selection of custom genes and desired cell types depends on the
user’s research interests and goals. It allows for flexibility in
focusing on specific genes and examining particular cell types of
interest in the visualization.</p>
</div>
<div class="section level3">
<h3 id="cell-type-specific-pairwise-correlation-analysis-and-visualization">Cell Type-specific Pairwise Correlation Analysis and
Visualization<a class="anchor" aria-label="anchor" href="#cell-type-specific-pairwise-correlation-analysis-and-visualization"></a>
</h3>
<p>This analysis aims to explore the correlation patterns between
different cell types in a single-cell gene expression dataset. The goal
is to compare the gene expression profiles of cells from a reference
dataset and a query dataset to understand the relationships and
similarities between various cell types.</p>
<p>To perform the analysis, we start by computing the pairwise
correlations between the query and reference cells for selected cell
types (“CD4”, “CD8”, “B_and_plasma”). The Spearman correlation method is
used, user can also use Pearsons correlation coeefficient.</p>
<p>This will return average correlation matrix which can be visulaized
by user’s method of choice. Here, the results are visualized as a
correlation plot using the corrplot package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeAveragePairwiseCorrelation.html">computeAveragePairwiseCorrelation</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                                    reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                                    query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                                    ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                                    cell_types <span class="op">=</span> <span class="va">selected_cell_types</span>, </span>
<span>                                                    correlation_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the pairwise average correlations using corrplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span>, method <span class="op">=</span> <span class="st">"number"</span>, tl.col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Cell-Type-Correlation-Analysis-Visualization-1.png" width="700"></p>
<p>In this case, users have the flexibility to extract the gene
expression profiles of specific cell types from the reference and query
datasets and provide these profiles as input to the function.
Additionally, they can select their own set of genes that they consider
relevant for computing the pairwise correlations. For demonstartion we
have used common highly variable genes from reference and query
dataset.</p>
<p>By providing their own gene expression profiles and choosing specific
genes, users can focus the analysis on the cell types and genes of
interest to their research question.</p>
</div>
<div class="section level3">
<h3 id="pairwise-distance-analysis-and-density-visualization">Pairwise Distance Analysis and Density Visualization<a class="anchor" aria-label="anchor" href="#pairwise-distance-analysis-and-density-visualization"></a>
</h3>
<p>This function serves to conduct a analysis of pairwise distances or
correlations between cells of specific cell types within a single-cell
gene expression dataset. By calculating these distances or correlations,
users can gain insights into the relationships and differences in gene
expression profiles between different cell types. The function
facilitates this analysis by generating density plots, allowing users to
visualize the distribution of distances or correlations for various
pairwise comparisons.</p>
<p>The analysis offers the flexibility to select a particular cell type
for examination, and users can choose between different distance
metrics, such as “euclidean” or “manhattan,” to calculate pairwise
distances.</p>
<p>To illustrate, the function is applied to the cell type CD8 using the
euclidean distance metric in the example below.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculatePairwiseDistancesAndPlotDensity.html">calculatePairwiseDistancesAndPlotDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                         reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                         query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                         ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                         cell_type_query <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         cell_type_reference <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         distance_metric <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Pairwise-Distance-Analysis-Density-Visualization-1.png" width="768"></p>
<p>Alternatively, users can opt for the “correlation” distance metric,
which measures the similarity in gene expression profiles between
cells.</p>
<p>To illustrate, the function is applied to the cell type CD8 using the
correlation distance metric in the example below. By selecting either
the “pearson” or “spearman” correlation method, users can emphasize
either linear or rank-based associations, respectively.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculatePairwiseDistancesAndPlotDensity.html">calculatePairwiseDistancesAndPlotDensity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                                         reference_data <span class="op">=</span> <span class="va">ref_data_subset</span>, </span>
<span>                                         query_cell_type_col <span class="op">=</span> <span class="st">"labels"</span>, </span>
<span>                                         ref_cell_type_col <span class="op">=</span> <span class="st">"reclustered.broad"</span>, </span>
<span>                                         cell_type_query <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         cell_type_reference <span class="op">=</span> <span class="st">"CD8"</span>, </span>
<span>                                         distance_metric <span class="op">=</span> <span class="st">"correlation"</span>,</span>
<span>                                         correlation_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Pairwise-Distance-Correlation-Based-Density-Visualization-1.png" width="768"></p>
<p>By utilizing this function, users can explore the pairwise distances
between query and reference cells of a specific cell type and gain
insights into the distribution of distances through density plots. This
analysis aids in understanding the similarities and differences in gene
expression profiles for the selected cell type within the query and
reference datasets.</p>
</div>
<div class="section level3">
<h3 id="pc-regression-analysis">PC regression analysis<a class="anchor" aria-label="anchor" href="#pc-regression-analysis"></a>
</h3>
<p>Performing PC regression analysis on a SingleCellExperiment object
enables users to examine the relationship between a principal component
(PC) from the dimension reduction slot and an independent variable of
interest. By specifying the desired dependent variable as one of the
principal components (e.g., “PC1”, “PC2”, etc.) and providing the
corresponding independent variable from the colData slot
(e.g. “cell_type”), users can explore the associations between linear
structure in the single-cell gene expression dataset (reference and
query) and an independent variable of interest (e.g. cell type or
batch).</p>
<p>The function prints two diagnostic plots by default:</p>
<ul>
<li>a plot of the two PCs with the highest R<sup>2</sup> with the
specified independent variable</li>
<li>a dot plot showing the R<sup>2</sup> of each consecutive PC ~
indep.var regression
<ul>
<li>Generally you should expect this plot to die off to near 0 before
~PC10</li>
<li>Interpretation example: If the R<sup>2</sup> values are high
(&gt;=50%) anywhere in PCs 1-5 and your independent variable is “batch”,
you have batch effects!</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Specify the dependent variables (principal components) and</span></span>
<span><span class="co">#  independent variable (e.g., "labels")</span></span>
<span><span class="va">dep.vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span></span>
<span><span class="va">indep.var</span> <span class="op">&lt;-</span> <span class="st">"labels"</span></span>
<span></span>
<span><span class="co"># Perform linear regression on multiple principal components</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>                    dep.vars <span class="op">=</span> <span class="va">dep.vars</span>, </span>
<span>                    indep.var <span class="op">=</span> <span class="va">indep.var</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/Regression-1.png" width="700"><img src="scDiagnostics_files/figure-html/Regression-2.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Print the summaries of the linear regression models and R-squared</span></span>
<span><span class="co">#  values</span></span>
<span></span>
<span><span class="co"># Summaries of the linear regression models</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">regression.summaries</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; lm(formula = f, data = df)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residuals:</span></span>
<span><span class="co">#&gt;      Min       1Q   Median       3Q      Max </span></span>
<span><span class="co">#&gt; -14.4118  -2.5919  -0.3286   2.5929  12.2589 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients:</span></span>
<span><span class="co">#&gt;                    Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">#&gt; (Intercept)         -9.6062     0.3500  -27.45   &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; IndependentCD4       7.7054     0.4108   18.76   &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; IndependentCD8      15.3986     0.4271   36.05   &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; IndependentMyeloid  11.9732     0.4753   25.19   &lt;2e-16 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residual standard error: 4.156 on 965 degrees of freedom</span></span>
<span><span class="co">#&gt; Multiple R-squared:  0.5996, Adjusted R-squared:  0.5984 </span></span>
<span><span class="co">#&gt; F-statistic: 481.7 on 3 and 965 DF,  p-value: &lt; 2.2e-16</span></span>
<span></span>
<span><span class="co"># R-squared values</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">rsquared</span></span>
<span><span class="co">#&gt;         PC1         PC2         PC3         PC4         PC5         PC6 </span></span>
<span><span class="co">#&gt; 0.599614123 0.537021405 0.435744738 0.139056106 0.117961322 0.114439358 </span></span>
<span><span class="co">#&gt;         PC7         PC8         PC9        PC10        PC11        PC12 </span></span>
<span><span class="co">#&gt; 0.039179544 0.001160446 0.002783929 0.005571931 0.001378947 0.002534793</span></span>
<span></span>
<span><span class="co"># Variance contributions for each principal component</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">var.contributions</span></span>
<span><span class="co">#&gt;          PC1          PC2          PC3          PC4          PC5          PC6 </span></span>
<span><span class="co">#&gt; 6.7286315831 2.7629320174 1.2945575966 0.3166828703 0.2377093993 0.1592056363 </span></span>
<span><span class="co">#&gt;          PC7          PC8          PC9         PC10         PC11         PC12 </span></span>
<span><span class="co">#&gt; 0.0426559772 0.0008730584 0.0020421974 0.0038728627 0.0008466853 0.0014487730</span></span>
<span></span>
<span><span class="co"># Total variance explained</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">total.variance.explained</span></span>
<span><span class="co">#&gt; [1] 11.55146</span></span></code></pre></div>
<p>This analysis helps uncover whether there is a systematic variation
in PC values across different cell types. In the example above, we can
see that the four cell types are spread out across both PC1 and PC2.
Digging into the genes with high loadins on these PCs can help explain
the biological or technical factors driving cellular heterogeneity. It
can help identify PC dimensions that capture variation specific to
certain cell types or distinguish different cellular states.</p>
<p>Let’s look at the genes driving PC1 by ordering the rotation matrix
by the absolute gene loadings for PC1:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc_df</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">)</span><span class="op">$</span><span class="va">PCA</span>, <span class="st">"rotation"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pc_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">pc_df</span><span class="op">$</span><span class="va">PC1</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;            PC1         PC2         PC3          PC4         PC5</span></span>
<span><span class="co">#&gt; GZMA 0.1517186  0.04586987 -0.08107421 -0.026938885 -0.02137103</span></span>
<span><span class="co">#&gt; GNLY 0.1552411 -0.08726472  0.12820759 -0.101178473 -0.31953772</span></span>
<span><span class="co">#&gt; CST7 0.1665665  0.02884329 -0.08880178 -0.034188322 -0.06706179</span></span>
<span><span class="co">#&gt; CCL4 0.1686632  0.03547543 -0.17317610 -0.009090169  0.01089660</span></span>
<span><span class="co">#&gt; CCL5 0.1815465  0.12081019 -0.19633461  0.008349436  0.06547467</span></span>
<span><span class="co">#&gt; NKG7 0.2286197  0.01800213 -0.04694001 -0.090383938 -0.13464307</span></span></code></pre></div>
<p>PC1 is mostly driven by NKG7 - Natural Killer Cell Granule Protein 7.
This gene is important in CD8+ T cells, so that makes sense that it’s
distinguishing the cell types shown.</p>
<blockquote>
<p>Exercise: What genes are driving PC2? Do they make sense?</p>
</blockquote>
<blockquote>
<p>Exercise: Try to use the command below to examine the spike on PC5.
What’s going on there?</p>
</blockquote>
<p><code>plotPCA(query_data, ncomponents = c(1,5), color_by = "labels")</code></p>
</div>
<div class="section level3">
<h3 id="outlier-detection">Outlier detection<a class="anchor" aria-label="anchor" href="#outlier-detection"></a>
</h3>
<p>Isolation forests are a fast and effective method for detecting
outliers in high-dimensional data without relying on density estimation.
<code><a href="../reference/calculateOutlierScore.html">calculateOutlierScore()</a></code> is a wrapper around the main
function of the <a href="https://cran.r-project.org/web/packages/isotree/index.html" class="external-link">isotree
package</a>. Here we run it on the PC embeddings but, it can also be
informative to run it on the raw counts. Here we can see some outliers
in PC1/2 space flagged by the default <code>score &gt; 0.5</code>
threshold due to their clear isolation (as well as a few that are
isolated on the non-visualized PCs).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateOutlierScore.html">calculateOutlierScore</a></span><span class="op">(</span><span class="va">query_data</span>, </span>
<span>                                    use_pcs <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                    plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in isotree::isolation.forest(X, output_score = TRUE, ...): Attempting</span></span>
<span><span class="co">#&gt; to use more than 1 thread, but package was compiled without OpenMP support. See</span></span>
<span><span class="co">#&gt; https://github.com/david-cortes/installing-optimized-libraries#4-macos-install-and-enable-openmp</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>Columns <code>outlier_score</code> and <code>is_outlier</code> are
added to the colData of the output. The isotree model object itself is
added to the metadata.</p>
<p>One can pass different <code>dimred</code> names to change the plot,
but running an isolation in forest in e.g. tSNE or UMAP space does not
make sense due to the non-linearity of these approaches. Hence this is
not allowed - when <code>use_pcs = FALSE</code>, the logcounts are used
instead.</p>
</div>
<div class="section level3">
<h3 id="annotation-entropy">Annotation entropy<a class="anchor" aria-label="anchor" href="#annotation-entropy"></a>
</h3>
<p>In order to assess the confidence of cell type predictions, we can
use the function <code><a href="../reference/calculateCategorizationEntropy.html">calculateCategorizationEntropy()</a></code>. This
function calculates the information entropy of assignment probabilities
across a set of cell types for each cell. If a set of class
probabilities are confident, the entropies will be low.</p>
<p>This can be used to compare two sets of cell type assignments
(e.g. from different type assignment methods) to compare their relative
confidence. <strong>Please note that this has nothing to do with their
accuracy!</strong> Computational methods can sometimes be confidently
incorrect.</p>
<p>The cell type probabilities should be passed as a matrix with cell
types as rows and cells as columns. If the columns of the matrix are not
valid probability distributions (i.e. don’t sum to 1 as in the below
example), the function will perform a column-wise softmax to convert
them to a probability scale. This may or may not work well depending on
the distribution of the inputs, so if at all possible try to pass
probabilities instead of arbitrary scores.</p>
<p>In this example, we create 500 random cells with random normal cell
type “scores” across 4 cell types. For demonstration we make the score
of the first class much higher in the first 250 cells. After the
softmax, this will equate to a very high probability of cell type 1. The
remaining 250 will have assignments that are roughly even across the
four cell types (i.e. high entropy).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">500</span> <span class="op">*</span> <span class="fl">4</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">X</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">250</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">250</span><span class="op">]</span> <span class="op">+</span> <span class="fl">5</span> </span>
<span></span>
<span><span class="va">entropy_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCategorizationEntropy.html">calculateCategorizationEntropy</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="co">#&gt; X doesn't seem to be on the probability scale,applying column-wise softmax.</span></span>
<span><span class="co">#&gt; Max possible entropy given 4 categories: 1.39</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>From the plot we can see that half of the cells (the first half we
shifted to class 1) have low entropy, and half have high entropy.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this analysis, we have demonstrated the capabilities of the
scDiagnostics package for assessing the appropriateness of cell
assignments in single-cell gene expression profiles. By utilizing
various diagnostic functions and visualization techniques, we have
explored different aspects of the data, including total UMI counts,
annotation scores, gene expression distributions, dimensional reduction
plots, gene set scores, pairwise correlations, pairwise distances, and
linear regression analysis.</p>
<p>Through the scatter plots, histograms, and dimensional reduction
plots, we were able to gain insights into the relationships between gene
expression patterns, cell types, and the distribution of cells in a
reduced-dimensional space. The examination of gene expression
distributions, gene sets, and pathways allowed us to explore the
functional landscape and identify subpopulations with distinct
characteristics within the dataset. Additionally, the pairwise
correlation and distance analyses provided a deeper understanding of the
similarities and differences between cell types, highlighting potential
relationships and patterns.</p>
<hr>
<div class="section level3">
<h3 id="r-session-info">R.session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h3>
<pre><code>R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Pacific/Honolulu
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] celldex_1.12.0              corrplot_0.92              
 [3] AUCell_1.24.0               SingleR_2.4.1              
 [5] RColorBrewer_1.1-3          scRNAseq_2.16.0            
 [7] scran_1.30.2                scater_1.30.1              
 [9] ggplot2_3.5.0               scuttle_1.12.0             
[11] scDiagnostics_0.99.0        SingleCellExperiment_1.24.0
[13] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[15] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
[17] IRanges_2.36.0              S4Vectors_0.40.2           
[19] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[21] matrixStats_1.3.0           BiocStyle_2.30.0           

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.3               
  [5] R.oo_1.26.0                   tibble_3.2.1                 
  [7] graph_1.80.0                  XML_3.99-0.16.1              
  [9] lifecycle_1.0.4               edgeR_4.0.16                 
 [11] lattice_0.22-6                ensembldb_2.26.0             
 [13] isotree_0.6.1-1               magrittr_2.0.3               
 [15] limma_3.58.1                  sass_0.4.9                   
 [17] rmarkdown_2.26                jquerylib_0.1.4              
 [19] yaml_2.3.8                    metapod_1.10.1               
 [21] httpuv_1.6.15                 DBI_1.2.2                    
 [23] abind_1.4-5                   zlibbioc_1.48.2              
 [25] purrr_1.0.2                   R.utils_2.12.3               
 [27] AnnotationFilter_1.26.0       RCurl_1.98-1.14              
 [29] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [31] ggrepel_0.9.5                 irlba_2.3.5.1                
 [33] annotate_1.80.0               dqrng_0.3.2                  
 [35] pkgdown_2.0.8                 DelayedMatrixStats_1.24.0    
 [37] codetools_0.2-20              DelayedArray_0.28.0          
 [39] xml2_1.3.6                    tidyselect_1.2.1             
 [41] farver_2.1.1                  ScaledMatrix_1.10.0          
 [43] viridis_0.6.5                 BiocFileCache_2.10.2         
 [45] GenomicAlignments_1.38.2      jsonlite_1.8.8               
 [47] BiocNeighbors_1.20.2          systemfonts_1.0.6            
 [49] tools_4.3.3                   progress_1.2.3               
 [51] ragg_1.3.0                    Rcpp_1.0.12                  
 [53] glue_1.7.0                    gridExtra_2.3                
 [55] SparseArray_1.2.4             xfun_0.43                    
 [57] dplyr_1.1.4                   withr_3.0.0                  
 [59] BiocManager_1.30.22           fastmap_1.1.1                
 [61] bluster_1.12.0                fansi_1.0.6                  
 [63] digest_0.6.35                 rsvd_1.0.5                   
 [65] R6_2.5.1                      mime_0.12                    
 [67] textshaping_0.3.7             colorspace_2.1-0             
 [69] biomaRt_2.58.2                RSQLite_2.3.6                
 [71] R.methodsS3_1.8.2             utf8_1.2.4                   
 [73] generics_0.1.3                data.table_1.15.4            
 [75] rtracklayer_1.62.0            prettyunits_1.2.0            
 [77] httr_1.4.7                    htmlwidgets_1.6.4            
 [79] S4Arrays_1.2.1                pkgconfig_2.0.3              
 [81] gtable_0.3.4                  blob_1.2.4                   
 [83] XVector_0.42.0                htmltools_0.5.8.1            
 [85] bookdown_0.38                 GSEABase_1.64.0              
 [87] ProtGenerics_1.34.0           scales_1.3.0                 
 [89] png_0.1-8                     knitr_1.46                   
 [91] rstudioapi_0.16.0             rjson_0.2.21                 
 [93] curl_5.2.1                    cachem_1.0.8                 
 [95] stringr_1.5.1                 BiocVersion_3.18.1           
 [97] parallel_4.3.3                vipor_0.4.7                  
 [99] AnnotationDbi_1.64.1          restfulr_0.0.15              
[101] desc_1.4.3                    pillar_1.9.0                 
[103] grid_4.3.3                    vctrs_0.6.5                  
[105] promises_1.3.0                BiocSingular_1.18.0          
[107] dbplyr_2.5.0                  beachmat_2.18.1              
[109] xtable_1.8-4                  cluster_2.1.6                
[111] beeswarm_0.4.0                evaluate_0.23                
[113] GenomicFeatures_1.54.4        cli_3.6.2                    
[115] locfit_1.5-9.9                compiler_4.3.3               
[117] Rsamtools_2.18.0              rlang_1.1.3                  
[119] crayon_1.5.2                  labeling_0.4.3               
[121] fs_1.6.3                      ggbeeswarm_0.7.2             
[123] stringi_1.8.3                 viridisLite_0.4.2            
[125] BiocParallel_1.36.0           munsell_0.5.1                
[127] Biostrings_2.70.3             lazyeval_0.2.2               
[129] Matrix_1.6-5                  ExperimentHub_2.10.0         
[131] hms_1.1.3                     sparseMatrixStats_1.14.0     
[133] bit64_4.0.5                   KEGGREST_1.42.0              
[135] statmod_1.5.0                 shiny_1.8.1.1                
[137] highr_0.10                    interactiveDisplayBase_1.40.0
[139] AnnotationHub_3.10.1          igraph_2.0.3                 
[141] memoise_2.0.1                 bslib_0.7.0                  
[143] bit_4.0.5                    </code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Smriti Chawla, Andrew Ghazi, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
