<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started with `scDiagnostics` • scDiagnostics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting Started with `scDiagnostics`">
<meta property="og:description" content="scDiagnostics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnostics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiagnostics.html">Geting Started with scDiagnostics</a>
    </li>
    <li>
      <a href="../articles/VisualizationTools.html">Visualization of Cell Type Annotations</a>
    </li>
    <li>
      <a href="../articles/QCandAnnotationScores.html">Visualization of QC and Annotation Scores</a>
    </li>
    <li>
      <a href="../articles/DatasetAlignment.html">Evaluation of Dataset and Marker Gene Alignment</a>
    </li>
    <li>
      <a href="../articles/StatisticalMeasures.html">Statistical Measures to Assess Dataset Alignment</a>
    </li>
    <li>
      <a href="../articles/AnnotationAnomalies.html">Detection of Annotation Anomalies</a>
    </li>
    <li>
      <a href="../articles/CellDistancesDiagnostics.html">Calculation of Distances Between Specific Cells and Cell Populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ccb-hms/scDiagnostics/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started with
<code>scDiagnostics</code>
</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><a class="author_email" href="mailto:#"></a><a href="mailto:anthony-alexander_christidis@hms.harvard.edu" class="email">anthony-alexander_christidis@hms.harvard.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Andrew
Ghazi</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Smriti
Chawla</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Nitesh
Turaga</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Ludwig
Geistlinger</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><h4 data-toc-skip class="author">Robert
Gentleman</h4>
            <address class="author_afil">
      Core for Computational Biomedicine, Harvard Medical
School<br><small class="dont-index">Source: <a href="https://github.com/ccb-hms/scDiagnostics/blob/HEAD/vignettes/scDiagnostics.Rmd" class="external-link"><code>vignettes/scDiagnostics.Rmd</code></a></small>
      <div class="hidden name"><code>scDiagnostics.Rmd</code></div>

    </address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h2>
<p>Annotation transfer from a reference dataset is a key process for
annotating cell types in new single-cell RNA-sequencing (scRNA-seq)
experiments. This approach provides a quick, automated, and reproducible
alternative to manual annotation based on marker gene expression.
Despite its advantages, challenges such as dataset imbalance and
unrecognized discrepancies between query and reference datasets can lead
to inaccurate annotations and affect subsequent analyses.</p>
<p>The <code>scDiagnostics</code> package is designed to address these
issues by offering a suite of diagnostic tools for the systematic
evaluation of cell type assignments in scRNA-seq data. It provides
functionality to assess whether query and reference datasets are
well-aligned, which is crucial for ensuring accurate annotation
transfer. In addition, <code>scDiagnostics</code> helps evaluate
annotation ambiguity, cluster heterogeneity, and marker gene alignment.
By providing insights into these aspects, <code>scDiagnostics</code>
enables researchers to determine the precision with which cells from a
new scRNA-seq experiment can be assigned to known cell types, thereby
supporting more accurate and reliable downstream analysis.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the development version of the package from Github, use
the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: you will need the <a href="https://cran.r-project.org/web/packages/remotes/index.html" class="external-link">remotes</a>
package to install from GitHub.</p>
<p>To build the package vignettes upon installation use:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ccb-hms/scDiagnostics"</span>,</span>
<span>                          build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Once you have installed the package, you can load it with the
following code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<p>To explore the full capabilities of the <code>scDiagnostics</code>
package, you have the option to use your own data or leverage the
datasets included within the scDiagnostics package itself. In this
guide, we will focus on utilizing these built-in datasets, which provide
a practical and convenient resource for demonstrating the features of
<code>scDiagnostics.</code> These datasets are specifically designed to
facilitate the exploration of the package’s functionalities and to help
evaluate the accuracy of cell type assignments. You can learn more about
the datasets by looking at the documentation of the datasets available
in the reference manual.</p>
<div class="section level3">
<h3 id="loading-datasets">Loading Datasets<a class="anchor" aria-label="anchor" href="#loading-datasets"></a>
</h3>
<p>In these datasets available in the <code>scDiagnostics</code>
package, <code>reference_data</code>, <code>query_data</code>, and
<code>qc_data</code> are all <code>SingleCellExperiment</code> objects
that include a <code>logcounts</code> assay, which stores the
log-transformed expression values for the genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"qc_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>The <code>reference_data</code> object is a curated dataset that has
been cleaned and processed, and it contains column data labeled
<code>expert_annotation</code>, which provides cell type annotations
assigned by experts. On the other hand, <code>query_data</code> also
includes <code>expert_annotation</code>, but it additionally features
<code>SingleR_annotation</code>, which is the cell type annotation
generated by the <code>SingleR</code> package, a popular package for
cell type assignment based on reference datasets. The
<code>qc_data</code> object contains a special column called
<code>annotation_scores</code>, which holds the scores from the
<code>SingleR</code> annotations, providing a measure of confidence or
relevance for the assigned cell types.</p>
<p>By working with these datasets, you can gain hands-on experience with
the various diagnostic tools and functions offered by scDiagnostics,
allowing you to better understand how well it aligns query and reference
datasets, assesses annotation ambiguity, and evaluates cluster
heterogeneity and marker gene alignment.</p>
</div>
<div class="section level3">
<h3 id="subsetting-the-datasets">Subsetting the Datasets<a class="anchor" aria-label="anchor" href="#subsetting-the-datasets"></a>
</h3>
<p>Some functions in the vignette are designed to work with
<code>SingleCellExperiment</code> objects that contain data from only
one cell type. We will create separate <code>SingleCellExperiment</code>
objects that only CD4 cells, to ensure compatibility with these
functions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset to CD4 cells</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Select highly variable genes</span></span>
<span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">ref_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_cd4</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data by common genes</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="va">ref_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="va">query_data_cd4</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Run PCA on both datasets</span></span>
<span><span class="va">ref_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">ref_data_cd4</span><span class="op">)</span></span>
<span><span class="va">query_data_cd4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_cd4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started-with-scdiagnostics">Getting Started with <code>scDiagnostics</code><a class="anchor" aria-label="anchor" href="#getting-started-with-scdiagnostics"></a>
</h2>
<p><strong>The functions introduced in this section represent just a
subset of the functions available in the <code>scDiagnostics</code>
package.</strong></p>
<p>For a complete overview and detailed demonstrations of all the
functions included in the package, please refer to the designated
vignettes. Each vignette is designed to address specific aspects of
<code>scDiagnostics</code>, and this vignette highlights key
functionalities to illustrate their applications. These vignettes
provide in-depth guidance and examples for each function, helping users
fully leverage the capabilities of <code>scDiagnostics</code> in their
single-cell analyses.</p>
</div>
<div class="section level2">
<h2 id="visualization-of-query-vs--reference-datasets">Visualization of Query vs. Reference Datasets<a class="anchor" aria-label="anchor" href="#visualization-of-query-vs--reference-datasets"></a>
</h2>
<p>For a detailed example of all possible functions to visualize
reference and query datasets, please refer to the “<strong>Visualization
of Cell Type Annotations</strong>” vignette.</p>
<div class="section level3">
<h3 id="plotcelltypepca">
<code>plotCellTypePCA</code><a class="anchor" aria-label="anchor" href="#plotcelltypepca"></a>
</h3>
<p>The <code>plotCellTypePCA</code> function provides a visual
comparison of principal components (PCs) for different cell types across
query and reference datasets. By projecting the query data onto the PCA
space of the reference dataset, it creates informative plots to help you
understand how various cell types are distributed in the principal
component space.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot PCA data</span></span>
<span><span class="va">pc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCellTypePCA.html">plotCellTypePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                           reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                           cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                           query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                           ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="co"># Display plot</span></span>
<span><span class="va">pc_plot</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;">
The function returns a <code>ggplot</code> object featuring pairwise
scatter plots of the selected principal components. Each plot compares
how different cell types from the query and reference datasets project
onto the PCA space. This visualization aids in identifying how cell
types distribute across PCs and facilitates comparisons between
datasets.</p>
<p>The <code>reference_data</code> argument contains the reference cell
data, which serves as the foundation for defining the PC space. The
<code>query_data</code> parameter includes the query cell data that will
be projected. The function uses the <code>ref_cell_type_col</code> and
<code>query_cell_type_col</code> to identify the relevant cell type
annotations in the reference and query datasets.</p>
</div>
<div class="section level3">
<h3 id="calculatediscriminantspace">
<code>calculateDiscriminantSpace</code><a class="anchor" aria-label="anchor" href="#calculatediscriminantspace"></a>
</h3>
<p>Alternatively, you can also use the
<code>calculateDiscriminantSpace</code> function, which projects query
single-cell RNA-seq data onto a discriminant space defined by a
reference dataset. This approach helps evaluate the similarity between
the query and reference data, offering insights into the classification
of query cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disc_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateDiscriminantSpace.html">calculateDiscriminantSpace</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                                          query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                          ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                                          query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span><span class="op">)</span></span></code></pre></div>
<p>The function returns a comprehensive output that includes
discriminant eigenvalues and eigenvectors, which represent the variance
explained by each discriminant axis and are used to project the data. It
also provides the projections of the reference and query data onto the
discriminant space. The Mahalanobis distances between the query and
reference cell types are calculated, offering insights into how close
the query projections are to the reference. The cosine similarity scores
provide another metric to assess the similarity between the
datasets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, plot_type <span class="op">=</span> <span class="st">"scatterplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-8-1.png" width="100%" style="display: block; margin: auto;">
Alternatively, you can create a boxplot comparing query and reference
projections for a specific cell type:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">disc_output</span>, cell_types <span class="op">=</span> <span class="st">"CD4-CD8"</span>, plot_type <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-9-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="visualization-of-marker-expressions">Visualization of Marker Expressions<a class="anchor" aria-label="anchor" href="#visualization-of-marker-expressions"></a>
</h3>
<p>Visualizing gene expression distributions is crucial for
understanding dataset similarity and cell type-specific expression
patterns. The <code>plotMarkerExpression</code> function allows you to
compare the expression levels of a specific gene between a reference
dataset and a query dataset, both overall and within a specified cell
type. This comparison is done using density plots, which help in
assessing the alignment and potential discrepancies between
datasets.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMarkerExpression.html">plotMarkerExpression</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                     query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                     ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                     query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                     gene_name <span class="op">=</span> <span class="st">"VPREB3"</span>, </span>
<span>                     cell_type <span class="op">=</span> <span class="st">"B_and_plasma"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-10-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-qc-and-annotation-scores">Visualization of QC and Annotation Scores<a class="anchor" aria-label="anchor" href="#visualization-of-qc-and-annotation-scores"></a>
</h2>
<p>For a detailed example of all possible functions to visualize quality
control and annotation score metrics, please refer to the
“<strong>Visualization of QC and Annotation Scores</strong>”
vignette.</p>
<p>Visualizing gene set scores on dimensional reduction plots helps in
understanding how gene set or pathway activity varies across single
cells in reduced dimensional spaces such as PCA, t-SNE, or UMAP. The
<code>plotGeneSetScores</code> function allows you to plot these scores
on the specified reduced dimensions, offering insights into the spatial
distribution of gene set activities in the context of single-cell
data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeneSetScores.html">plotGeneSetScores</a></span><span class="op">(</span>se_object <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                  score_col <span class="op">=</span> <span class="st">"gene_set_scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-11-1.png" width="100%" style="display: block; margin: auto;">
The function returns a <code>ggplot</code> object displaying scatter
plots of gene set or pathway scores on the specified reduced dimensions.
For PCA, the plot includes the percentage of variance explained for each
principal component in the legend.</p>
</div>
<div class="section level2">
<h2 id="evaluation-of-dataset-and-marker-gene-alignment">Evaluation of Dataset and Marker Gene Alignment<a class="anchor" aria-label="anchor" href="#evaluation-of-dataset-and-marker-gene-alignment"></a>
</h2>
<p>For a detailed example of all possible functions to assess the
alignment of datasets and marker genes, please refer to the
“<strong>Evaluation of Dataset and Marker Gene Alignment</strong>”
vignette.</p>
<div class="section level3">
<h3 id="comparecca">
<code>compareCCA</code><a class="anchor" aria-label="anchor" href="#comparecca"></a>
</h3>
<p>The <code>compareCCA</code> function is designed to compare the
subspaces spanned by the top principal components (PCs) of two datasets
using Canonical Correlation Analysis (CCA). The function calculates
canonical variables, correlations, and a similarity measure between the
subspaces of the reference and query datasets. This can be useful in
determining the similarity of cell populations or other high-dimensional
data structures across different conditions or datasets.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform CCA comparison</span></span>
<span><span class="va">cca_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compareCCA.html">compareCCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="co"># Visualize results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cca_comparison</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-12-1.png" width="100%" style="display: block; margin: auto;">
This function is useful for identifying similarities and differences in
cellular subpopulations or other high-dimensional datasets across
different conditions. The combination of CCA with PCA provides a
reliable framework for subspace comparison.</p>
</div>
<div class="section level3">
<h3 id="comparepca">
<code>comparePCA</code><a class="anchor" aria-label="anchor" href="#comparepca"></a>
</h3>
<p>The comparePCA function allows for the comparison of principal
components (PCs) obtained from separate PCA analyses on reference and
query datasets for a specific cell type. This comparison can be
performed using either cosine similarity or correlation metrics,
offering insights into the relationship between the subspaces spanned by
the PCs of the two datasets.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform CCA comparison</span></span>
<span><span class="va">similarity_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparePCA.html">comparePCA</a></span><span class="op">(</span></span>
<span>    query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>    reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>,</span>
<span>    query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="co"># Visualize the similarity matrix as a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">similarity_matrix</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-13-1.png" width="100%" style="display: block; margin: auto;">
The resulting similarity matrix can be visualized as a heatmap using the
plot function. This heatmap provides a clear graphical representation of
how similar or different the PCs of the reference and query datasets
are, based on the chosen metric.</p>
</div>
<div class="section level3">
<h3 id="plotwassersteindistance">
<code>plotWassersteinDistance</code><a class="anchor" aria-label="anchor" href="#plotwassersteindistance"></a>
</h3>
<p>The code below illustrates how to use the
<code>plotWassersteinDistance</code> function to compare the Wasserstein
distances between CD4 cells in the reference and query datasets. The
resulting plot provides insight into whether the differences between the
datasets are statistically significant.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the Wasserstein distance density plot</span></span>
<span><span class="fu"><a href="../reference/plotWassersteinDistance.html">plotWassersteinDistance</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_cd4</span>,</span>
<span>                        reference_data <span class="op">=</span> <span class="va">ref_data_cd4</span>, </span>
<span>                        query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                        pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                        alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-14-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="calculatevarimpoverlap">
<code>calculateVarImpOverlap</code><a class="anchor" aria-label="anchor" href="#calculatevarimpoverlap"></a>
</h3>
<p>Imagine you have two sets of data: one called reference_data and
another called query_data. Both sets include information about gene
expression and cell types, with columns named
<code>expert_annotation</code> for the reference data and
<code>SingleR_annotation</code> for the query data. You want to find out
which genes are most important for each dataset and then compare
them.</p>
<p>Here’s how you can do it with the function:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RF function to compare (between datasets) which genes are best at differentiating cell types</span></span>
<span><span class="va">rf_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateVarImpOverlap.html">calculateVarImpOverlap</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                    n_top <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Comparison table</span></span>
<span><span class="va">rf_output</span><span class="op">$</span><span class="va">var_imp_comparison</span></span>
<span><span class="co">#&gt;     B_and_plasma-CD4     B_and_plasma-CD8 B_and_plasma-Myeloid </span></span>
<span><span class="co">#&gt;                 0.82                 0.72                 0.34 </span></span>
<span><span class="co">#&gt;              CD4-CD8          CD4-Myeloid          CD8-Myeloid </span></span>
<span><span class="co">#&gt;                 0.72                 0.42                 0.24</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="statistical-measures-to-assess-dataset-alignment">Statistical Measures to Assess Dataset Alignment<a class="anchor" aria-label="anchor" href="#statistical-measures-to-assess-dataset-alignment"></a>
</h2>
<p>For a detailed example of all possible functions that implement
statistical methods to assess cell type annotation, please refer to the
“<strong>Statistical Measures to Assess Dataset Alignment</strong>”
vignette.</p>
<div class="section level3">
<h3 id="calculateaveragepairwisecorrelation">
<code>calculateAveragePairwiseCorrelation</code><a class="anchor" aria-label="anchor" href="#calculateaveragepairwisecorrelation"></a>
</h3>
<p>The <code>calculateAveragePairwiseCorrelation</code> function
computes the average pairwise correlations between specified cell types
in single-cell gene expression data. It calculates pairwise correlations
between query and reference cells using a specified correlation method,
and then averages these correlations for each cell type pair. This
approach helps assess the similarity between cells in reference and
query datasets and provides insights into the reliability of cell type
annotations.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the average pairwise correlations</span></span>
<span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateAveragePairwiseCorrelation.html">calculateAveragePairwiseCorrelation</a></span><span class="op">(</span></span>
<span>  query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>  reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>  query_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>  cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span>
<span><span class="co"># Visualize the correlation matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-16-1.png" width="100%" style="display: block; margin: auto;">
The displays a heatmap where each cell type pair is represented with a
color indicating the average correlation value between them. This
visualization helps in assessing the similarity between cell types
across the query and reference datasets.</p>
</div>
<div class="section level3">
<h3 id="regresspc">
<code>regressPC</code><a class="anchor" aria-label="anchor" href="#regresspc"></a>
</h3>
<p>The <code>regressPC</code> function performs linear regression of a
covariate of interest onto one or more principal components (PCs) based
on the data contained within a <code>SingleCellExperiment</code> object.
This method is useful for quantifying the variance explained by a
covariate, which can be applied to various aspects of single-cell
analysis, such as evaluating batch effects, clustering homogeneity, or
aligning query and reference datasets in cell type annotations.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PC regression with query data</span></span>
<span><span class="va">regress_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regressPC.html">regressPC</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                         query_data <span class="op">=</span> <span class="va">query_data</span>,</span>
<span>                         ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                         query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                         cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                         pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></span>
<span><span class="co"># Plot the results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"r_squared"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-17-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">regress_res</span>, plot_type <span class="op">=</span> <span class="st">"p-value"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-17-2.png" width="100%" style="display: block; margin: auto;"></p>
<p>The function calculates how much variance in the data matrix is
explained by a covariate through principal component regression. It
computes the R-squared value for each principal component and derives
the variance contribution of the covariate for each principal
component.</p>
</div>
</div>
<div class="section level2">
<h2 id="detection-of-annotation-anomalies">Detection of Annotation Anomalies<a class="anchor" aria-label="anchor" href="#detection-of-annotation-anomalies"></a>
</h2>
<p>For a detailed example of all possible functions that implement
statistical methods to assess cell type annotation, please refer to the
“<strong>Detection of Annotation Anomalies</strong>” vignette.</p>
<p>The <code>detectAnomaly</code> function is designed to identify
anomalies in single-cell data by leveraging PCA projections and the
Isolation Forest algorithm. This method is useful for detecting
anomalies or unusual patterns in single-cell datasets, whether you’re
analyzing a reference dataset or comparing a query dataset against
it.</p>
<p>The function projects single-cell data onto a PCA space and builds an
Isolation Forest model on this PCA space to detect anomalies. If a query
dataset is provided, the function computes anomaly scores for the query
data based on its PCA projections relative to the reference data. If no
query data is provided, it computes anomaly scores for the reference
data itself.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform anomaly detection</span></span>
<span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span><span class="op">)</span></span>
<span><span class="co"># Plot the results for a specific cell type</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anomaly_output</span>, </span>
<span>     cell_type <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>     data_type <span class="op">=</span> <span class="st">"query"</span>,</span>
<span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-18-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="calculation-of-distances-between-specific-cells-and-cell-types">Calculation of Distances Between Specific Cells and Cell Types<a class="anchor" aria-label="anchor" href="#calculation-of-distances-between-specific-cells-and-cell-types"></a>
</h2>
<p>For a detailed example of all possible functions to analyze distances
between specific cells and cell distributions, please refer to the
“<strong>Calculation of Distances Between Specific Cells and Cell
Types</strong>” vignette.</p>
<p>The <code>calculateCellDistances</code> function computes distances
both within a reference dataset and between query cells and reference
cells for each specified cell type. By first projecting data onto a PCA
space, the function calculates Euclidean distances to quantify
similarities and dissimilarities. For each cell type, it generates
pairwise distances within the reference dataset and measures how far
each query cell is from all reference cells. This approach enables
detailed analysis of cell type-specific distances, aiding in the
identification of outliers and other patterns of interest.</p>
<p>To identify anomalous cells within the query data, we first use the
<code>detectAnomaly</code> function, focusing specifically on the CD4
cell type. This function will compute anomaly scores for each CD4 cell
in the query dataset based on their projection in the PCA space of the
reference data. Next, we will plot the distance distributions for the
top 6 CD4 cells with the highest anomaly scores. These distances,
computed using the <code>calculateCellDistances</code> function, will
illustrate how these anomalous cells differ from the reference cells,
providing insight into their potential outlier status and helping to
visualize patterns within the data.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify outliers for CD4</span></span>
<span><span class="va">cd4_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectAnomaly.html">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                               query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                               query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                               ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="va">cd4_top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">cd4_anomalies</span><span class="op">$</span><span class="va">CD4</span><span class="op">$</span><span class="va">query_anomaly_scores</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PC data</span></span>
<span><span class="va">distance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateCellDistances.html">calculateCellDistances</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                        reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                        query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                        ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the densities of the distances</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD4"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiagnostics_files/figure-html/unnamed-chunk-19-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette, we have introduced the core functionalities of the
scDiagnostics package, providing a detailed guide on its use for
scRNA-seq data analysis. We started by exploring the basic capabilities,
including the visualization of query versus reference datasets through
dimensionality reduction techniques. This approach allows for both the
evaluation of multiple cell types simultaneously and the comparison of
distributions for individual cell types.</p>
<p>We then delved into the visualization of annotation scores, which is
essential for assessing the accuracy of cell type assignments. Following
this, we discussed statistical methods for evaluating annotation
precision. This includes assessing correlation between cell types and
performing statistical tests to compare two datasets, offering a
rigorous approach to validate cell type classifications.</p>
<p>The vignette also covered techniques for detecting annotation
anomalies, crucial for identifying and addressing potential issues that
could impact data validity. Additionally, we explored cell distance
diagnostics, focusing on analyzing distances between specific cells and
various cell populations or distributions to better understand internal
relationships and variations.</p>
<p>Overall, this guide equips you with practical tools and methods to
enhance the reliability and accuracy of your scRNA-seq data analysis
using scDiagnostics, paving the way for more informed and meaningful
biological interpretations.</p>
<hr>
</div>
<div class="section level2">
<h2 id="r-session-info">R Session Info<a class="anchor" aria-label="anchor" href="#r-session-info"></a>
</h2>
<pre><code>R version 4.4.0 (2024-04-24 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.32.0               ggplot2_3.5.1              
 [3] scran_1.32.0                scuttle_1.14.0             
 [5] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
 [7] Biobase_2.64.0              GenomicRanges_1.56.0       
 [9] GenomeInfoDb_1.40.1         IRanges_2.38.0             
[11] S4Vectors_0.42.0            BiocGenerics_0.50.0        
[13] MatrixGenerics_1.16.0       matrixStats_1.3.0          
[15] scDiagnostics_0.99.6        BiocStyle_2.32.0           

loaded via a namespace (and not attached):
 [1] DBI_1.2.2                 gridExtra_2.3            
 [3] rlang_1.1.3               magrittr_2.0.3           
 [5] compiler_4.4.0            DelayedMatrixStats_1.26.0
 [7] systemfonts_1.1.0         vctrs_0.6.5              
 [9] pkgconfig_2.0.3           crayon_1.5.2             
[11] fastmap_1.2.0             XVector_0.44.0           
[13] labeling_0.4.3            utf8_1.2.4               
[15] rmarkdown_2.27            UCSC.utils_1.0.0         
[17] ggbeeswarm_0.7.2          ragg_1.3.2               
[19] purrr_1.0.2               xfun_0.44                
[21] bluster_1.14.0            zlibbioc_1.50.0          
[23] cachem_1.1.0              beachmat_2.20.0          
[25] jsonlite_1.8.8            highr_0.11               
[27] DelayedArray_0.30.1       BiocParallel_1.38.0      
[29] irlba_2.3.5.1             parallel_4.4.0           
[31] cluster_2.1.6             biglm_0.9-3              
[33] R6_2.5.1                  bslib_0.7.0              
[35] ranger_0.16.0             limma_3.60.2             
[37] jquerylib_0.1.4           Rcpp_1.0.12              
[39] bookdown_0.39             knitr_1.46               
[41] Matrix_1.7-0              igraph_2.0.3             
[43] tidyselect_1.2.1          rstudioapi_0.16.0        
[45] abind_1.4-5               yaml_2.3.8               
[47] viridis_0.6.5             codetools_0.2-20         
[49] lattice_0.22-6            tibble_3.2.1             
[51] withr_3.0.0               evaluate_0.23            
[53] desc_1.4.3                pillar_1.9.0             
[55] BiocManager_1.30.23       generics_0.1.3           
[57] sparseMatrixStats_1.16.0  munsell_0.5.1            
[59] scales_1.3.0              glue_1.7.0               
[61] metapod_1.12.0            tools_4.4.0              
[63] speedglm_0.3-5            data.table_1.15.4        
[65] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0      
[67] locfit_1.5-9.9            fs_1.6.4                 
[69] grid_4.4.0                edgeR_4.2.0              
[71] colorspace_2.1-0          GenomeInfoDbData_1.2.12  
[73] beeswarm_0.4.0            BiocSingular_1.20.0      
[75] vipor_0.4.7               cli_3.6.2                
[77] rsvd_1.0.5                textshaping_0.4.0        
[79] fansi_1.0.6               viridisLite_0.4.2        
[81] S4Arrays_1.4.1            dplyr_1.1.4              
[83] gtable_0.3.5              isotree_0.6.1-1          
[85] sass_0.4.9                digest_0.6.35            
[87] SparseArray_1.4.8         ggrepel_0.9.5            
[89] dqrng_0.4.1               farver_2.1.2             
[91] htmlwidgets_1.6.4         memoise_2.0.1            
[93] htmltools_0.5.8.1         pkgdown_2.0.9            
[95] lifecycle_1.0.4           httr_1.4.7               
[97] transport_0.15-2          statmod_1.5.0            
[99] MASS_7.3-60.2            </code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
